% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/Get_and_Filter_CpGs.R
\name{getCpGs}
\alias{getCpGs}
\title{Read Bismark CpG reports}
\usage{
getCpGs(
  colData,
  path = getwd(),
  pattern = "*CpG_report.txt.gz",
  sameLoci = TRUE,
  chroms = c(paste("chr", 1:22, sep = ""), "chrX", "chrY", "chrM"),
  BPPARAM = BiocParallel::MulticoreParam(10),
  save = TRUE,
  file = "Unfiltered_BSseq.rds",
  verbose = TRUE
)
}
\arguments{
\item{colData}{A \code{data.frame} whose row names specify CpG reports to
load into the \code{\link[bsseq:`BSseq-class`]{BSseq}} object and
whose columns are sample traits with numeric values.}

\item{path}{A \code{character} giving the path(s) to the CpG reports.}

\item{pattern}{A regular expression used to filter for CpG reports.}

\item{sameLoci}{A \code{logical(1)} indicating whether CpG reports contain
the same set of methylation loci. This is the case if the files are
genome wide cytosine reports aligned to the same reference genome.
The default \code{sameLoci = TRUE} speeds up \code{getCpGs()} by
only having to parse each CpG report once.}

\item{chroms}{A \code{character} giving the chromosomes to include in the
\code{\link[bsseq:`BSseq-class`]{BSseq}} object.}

\item{BPPARAM}{A \code{\link[BiocParallel]{BiocParallelParam}} instance
providing the parallel back-end to use during evaluation.}

\item{save}{A \code{logical(1)} indicating whether to save the
\code{\link[bsseq:`BSseq-class`]{BSseq}} object.}

\item{file}{A \code{character(1)} giving the file name (.rds) for the saved
\code{\link[bsseq:`BSseq-class`]{BSseq}} object.}

\item{verbose}{A \code{logical(1)} indicating whether messages should be
printed.}
}
\value{
A \code{\link[bsseq:`BSseq-class`]{BSseq}} object.
}
\description{
\code{getCpGs()} reads individual sample Bismark CpG reports into a single
\code{\link[bsseq:`BSseq-class`]{BSseq}} object and then saves it as a .rds
file.
}
\details{
This \code{\link[bsseq:`BSseq-class`]{BSseq}} object still needs to be
filtered for coverage at individual CpGs. More information on these arguments
is given in the documentation for \code{\link[bsseq]{read.bismark()}}.
}
\examples{
\dontrun{

# Read Bismark CpG Reports
colData <- read.xlsx("sample_info.xlsx", rowNames = TRUE)
bs <- getCpGs(colData, file = "Unfiltered_BSseq.rds")

# Examine CpG Totals at Different Cutoffs
CpGtotals <- getCpGtotals(bs, file = "CpG_Totals.txt")
plotCpGtotals(CpGtotals, file = "CpG_Totals.pdf")

# Filter BSseq Object
bs <- filterCpGs(bs, cov = 2, perSample = 0.75, file = "Filtered_BSseq.rds")
}

}
\seealso{
\itemize{
\item \code{\link{getCpGtotals()}} and \code{\link{plotCpGtotals()}}
for help with deciding coverage cutoffs.
\item \code{\link{filterCpGs()}} to filter the
\code{\link[bsseq:`BSseq-class`]{BSseq}} object.
\item \code{\link[bsseq]{read.bismark()}} for more details on the
arguments and the underlying functions.
}
}
