<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="comethyl">
<title>Gene Body Analysis • comethyl</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Gene Body Analysis">
<meta property="og:description" content="comethyl">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">comethyl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/comethyl.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cemordaunt/comethyl/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Gene Body Analysis</h1>
                        <h4 data-toc-skip class="author">Charles Mordaunt and Julia Mouat</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cemordaunt/comethyl/blob/HEAD/vignettes/Gene_Body_Analysis.Rmd" class="external-link"><code>vignettes/Gene_Body_Analysis.Rmd</code></a></small>
      <div class="d-none name"><code>Gene_Body_Analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we use Comethyl to construct a weighted region comethylation network from WGBS data using CpGs grouped by gene body annotation. Because gene body methylation has been previously shown to correlate with expression, this approach allows for more direct gene mapping and functional interpretation. We identify modules of comethylated regions, investigate correlations with sample traits, and analyze functional enrichments.</p>
<p>The data set includes 74 male cord blood samples from newborns who were later diagnosed with autism spectrum disorder (ASD) and those with typical development (TD). Comethylation modules were associated with 49 sample characteristics including diagnosis, cell types, sample sequencing information such as percent CpG methylation, and demographic data such as home ownership. The goal of this analysis is to explore interactions between the methylome and sample traits prior to diagnosis with ASD.</p>
<p>Raw data is available on GEO (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140730" class="external-link">GSE140730</a>), see the <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00785-8" class="external-link">previous publication</a> for more details.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<pre><code><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cemordaunt/comethyl" class="external-link">comethyl</a></span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="set-global-options">Set Global Options<a class="anchor" aria-label="anchor" href="#set-global-options"></a>
</h2>
<p><code>enableWGCNAthreads()</code> allows parallel calculations with the specified number of threads. If the number of threads is not provided, the default is the number of processors online.</p>
<pre><code><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>R_THREADS <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu">enableWGCNAThreads</span><span class="op">(</span>nThreads <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="read-bismark-cpg-reports">Read Bismark CpG Reports<a class="anchor" aria-label="anchor" href="#read-bismark-cpg-reports"></a>
</h2>
<p>We read in an excel table with <code><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">openxlsx::read.xlsx()</a></code> where the first column includes the names of sample Bismark CpG reports and all other columns include trait values for each sample. All trait values must be numeric, though traits can be categorical or continuous. <code><a href="../reference/getCpGs.html">getCpGs()</a></code> reads individual sample Bismark CpG reports into a single BSseq object and then saves it as a .rds file. See <a href="https://cemordaunt.github.io/comethyl/articles/comethyl.html#inputs">Inputs</a> for more information.</p>
<pre><code><span class="va">colData</span> <span class="op">&lt;-</span> <span class="fu">read.xlsx</span><span class="op">(</span><span class="st">"sample_info.xlsx"</span>, rowNames <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCpGs.html">getCpGs</a></span><span class="op">(</span><span class="va">colData</span>, file <span class="op">=</span> <span class="st">"Unfiltered_BSseq.rds"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="examine-cpg-totals-at-different-cutoffs">Examine CpG Totals at Different Cutoffs<a class="anchor" aria-label="anchor" href="#examine-cpg-totals-at-different-cutoffs"></a>
</h2>
<p><code><a href="../reference/getCpGtotals.html">getCpGtotals()</a></code> calculates the total number and percent of CpGs remaining in a BSseq object after filtering at different <code>cov</code> (coverage) and <code>perSample</code> cutoffs and then saves it as a tab-separated text file. The purpose of this function is to help determine cutoffs to maximize the number of CpGs with sufficient data after filtering. Typically, the number of CpGs covered in 100% of samples decreases as the sample size increases, especially with low-coverage datasets. The goal for filtering is to try and balance the sequencing depth per CpG and the number of samples with the total number of CpGs.</p>
<p><code><a href="../reference/plotCpGtotals.html">plotCpGtotals()</a></code> plots the number of CpGs remaining after filtering by different combinations of <code>cov</code> and <code>perSample</code> in a line plot and then saves it as a PDF. <code><a href="../reference/plotCpGtotals.html">plotCpGtotals()</a></code> is designed to be used in combination with <code><a href="../reference/getCpGtotals.html">getCpGtotals()</a></code>. A ggplot is produced and can be further edited outside of this function if desired.</p>
<pre><code><span class="va">CpGtotals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCpGtotals.html">getCpGtotals</a></span><span class="op">(</span><span class="va">bs</span>, file <span class="op">=</span> <span class="st">"CpG_Totals.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotCpGtotals.html">plotCpGtotals</a></span><span class="op">(</span><span class="va">CpGtotals</span>, file <span class="op">=</span> <span class="st">"CpG_Totals.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="CpG%20Cluster%20Analysis/CpG_Totals.png" alt=""><p class="caption">Figure 1. CpG Totals</p>
</div>
</div>
<div class="section level2">
<h2 id="filter-bsobject">Filter BSobject<a class="anchor" aria-label="anchor" href="#filter-bsobject"></a>
</h2>
<p><code><a href="../reference/filterCpGs.html">filterCpGs()</a></code> subsets a BSseq object to include only those CpGs meeting <code>cov</code> and <code>perSample</code> cutoffs and then saves it as a .rds file. <code><a href="../reference/filterCpGs.html">filterCpGs()</a></code> is designed to be used after <code>cov</code> and <code>perSample</code> arguments have been optimized by <code><a href="../reference/getCpGtotals.html">getCpGtotals()</a></code> and <code><a href="../reference/plotCpGtotals.html">plotCpGtotals()</a></code>. Here we keep only CpGs with at least 2 reads in at least 75% of samples.</p>
<pre><code><span class="va">bs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterCpGs.html">filterCpGs</a></span><span class="op">(</span><span class="va">bs</span>, cov <span class="op">=</span> <span class="fl">2</span>, perSample <span class="op">=</span> <span class="fl">0.75</span>, file <span class="op">=</span> <span class="st">"Filtered_BSseq.rds"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="call-regions">Call Regions<a class="anchor" aria-label="anchor" href="#call-regions"></a>
</h2>
<p><code><a href="../reference/getRegions.html">getRegions()</a></code> generates a set of regions and some statistics based on the CpGs in a BSseq object. Regions can be defined based on CpG locations, built-in genomic annotations from <code>annotatr</code> (as here for gene bodies), or a custom genomic annotation.</p>
<p><code><a href="../reference/plotRegionStats.html">plotRegionStats()</a></code> plots histograms of region statistics, while <code><a href="../reference/plotSDstats.html">plotSDstats()</a></code> plots methylation standard deviation versus region statistics. With these plots, we can get an idea of the characteristics of our regions and see how methylation variability is affected. The goal is to identify regions with biological variability rather than technical variability (due to low coverage).</p>
<pre><code><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegions.html">getRegions</a></span><span class="op">(</span><span class="va">bs</span>, annotation <span class="op">=</span> <span class="st">"genes"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, file <span class="op">=</span> <span class="st">"Unfiltered_Regions.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionStats.html">plotRegionStats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>, file <span class="op">=</span> <span class="st">"Unfiltered_Region_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Unfiltered_Region_Plots.png" alt=""><p class="caption">Figure 2. Unfiltered Region Plots</p>
</div>
<pre><code><span class="fu"><a href="../reference/plotSDstats.html">plotSDstats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>, file <span class="op">=</span> <span class="st">"Unfiltered_SD_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Unfiltered_SD_Plots.png" alt=""><p class="caption">Figure 3. Unfiltered SD Plots</p>
</div>
</div>
<div class="section level2">
<h2 id="examine-region-totals-at-different-cutoffs">Examine Region Totals at Different Cutoffs<a class="anchor" aria-label="anchor" href="#examine-region-totals-at-different-cutoffs"></a>
</h2>
<p><code><a href="../reference/getRegionTotals.html">getRegionTotals()</a></code> calculates region totals at specified covMin and methSD cutoffs. Total regions (and thus total width and CpGs) are expected to decrease as the minimum coverage cutoff increases and SD cutoff increases. Here we also adjust the potential covMin and methSD cutoffs to match the higher coverage and lower methylation variation of gene bodies. <code><a href="../reference/plotRegionTotals.html">plotRegionTotals()</a></code> plots these region totals by covMin and methSD cutoffs.</p>
<pre><code><span class="va">regionTotals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegionTotals.html">getRegionTotals</a></span><span class="op">(</span><span class="va">regions</span>, covMin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span>,<span class="fl">10</span><span class="op">)</span>, methSD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.02</span>,<span class="fl">0.002</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"Region_Totals.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionTotals.html">plotRegionTotals</a></span><span class="op">(</span><span class="va">regionTotals</span>, file <span class="op">=</span> <span class="st">"Region_Totals.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Region_Totals.png" alt=""><p class="caption">Figure 4. Region Totals</p>
</div>
</div>
<div class="section level2">
<h2 id="filter-regions">Filter Regions<a class="anchor" aria-label="anchor" href="#filter-regions"></a>
</h2>
<p><code><a href="../reference/filterRegions.html">filterRegions()</a></code> subsets the regions to only include those meeting <code>covMin</code> and <code>methSD</code> cutoffs. <code><a href="../reference/filterRegions.html">filterRegions()</a></code> is designed to be used after <code>covMin</code> and <code>methSD</code> functions have been optimized with <code><a href="../reference/getRegionTotals.html">getRegionTotals()</a></code> and <code><a href="../reference/plotRegionTotals.html">plotRegionTotals()</a></code>. Here we filter for regions with at least 10 reads in all samples. We don’t use a standard deviation filter here because the region set is small enough already to use for network construction. Next, we examine our regions again with <code><a href="../reference/plotRegionStats.html">plotRegionStats()</a></code> after filtering.</p>
<pre><code><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterRegions.html">filterRegions</a></span><span class="op">(</span><span class="va">regions</span>, covMin <span class="op">=</span> <span class="fl">10</span>, methSD <span class="op">=</span> <span class="fl">0</span>, file <span class="op">=</span> <span class="st">"Filtered_Regions.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionStats.html">plotRegionStats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>, file <span class="op">=</span> <span class="st">"Filtered_Region_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Filtered_Region_Plots.png" alt=""><p class="caption">Figure 5. Filtered Region Plots</p>
</div>
</div>
<div class="section level2">
<h2 id="adjust-methylation-data-for-principal-components">Adjust Methylation Data for Principal Components<a class="anchor" aria-label="anchor" href="#adjust-methylation-data-for-principal-components"></a>
</h2>
<p><code><a href="../reference/getRegionMeth.html">getRegionMeth()</a></code> calculates region methylation from a BSseq object and saves it as a .rds. <code><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix()</a></code> creates a design matrix for our set of samples. <code><a href="../reference/adjustRegionMeth.html">adjustRegionMeth()</a></code> adjusts the region methylation for the top principal components and saves it as a .rds file. <code><a href="../reference/getDendro.html">getDendro()</a></code> clusters the samples based on the adjusted region methylation using Euclidean distance, while <code><a href="../reference/plotDendro.html">plotDendro()</a></code> plots the dendrogram. We can use this dendrogram to see if there are any outlier samples or samples clustering separately due to batch effects.</p>
<pre><code><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegionMeth.html">getRegionMeth</a></span><span class="op">(</span><span class="va">regions</span>, bs <span class="op">=</span> <span class="va">bs</span>, file <span class="op">=</span> <span class="st">"Region_Methylation.rds"</span><span class="op">)</span>
<span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">1</span>, data <span class="op">=</span> <span class="fu">pData</span><span class="op">(</span><span class="va">bs</span><span class="op">)</span><span class="op">)</span>
<span class="va">methAdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRegionMeth.html">adjustRegionMeth</a></span><span class="op">(</span><span class="va">meth</span>, mod <span class="op">=</span> <span class="va">mod</span>, file <span class="op">=</span> <span class="st">"Adjusted_Region_Methylation.rds"</span><span class="op">)</span>
<span class="fu"><a href="../reference/getDendro.html">getDendro</a></span><span class="op">(</span><span class="va">methAdj</span>, distance <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
<span class="fu"><a href="../reference/plotDendro.html">plotDendro</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"Sample_Dendrogram.pdf"</span>, expandY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>,<span class="fl">0.08</span><span class="op">)</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_Dendrogram.png" alt=""><p class="caption">Figure 6. Sample Dendrogram</p>
</div>
</div>
<div class="section level2">
<h2 id="select-soft-power-threshold">Select Soft Power Threshold<a class="anchor" aria-label="anchor" href="#select-soft-power-threshold"></a>
</h2>
<p><code><a href="../reference/getSoftPower.html">getSoftPower()</a></code> analyzes scale-free topology with Pearson or Bicor correlations to determine the best soft-thresholding power. This refers to the power to which all correlations are raised and how much more stronger correlations are weighted compared to weaker correlations. Pearson correlation is more sensitive than Bicor correlation, but is also more influenced by outlier samples. We use Pearson correlation in order to have higher power to detect correlated regions in a dataset with relatively low variability between samples.</p>
<p><code><a href="../reference/plotSoftPower.html">plotSoftPower()</a></code> plots the soft power threshold against scale free topology fit and connectivity. Typically, as the soft power threshold increases, fit increases and connectivity decreases. A soft power threshold should be selected as the lowest threshold where fit is 0.8 or higher (here we use 12).</p>
<pre><code><span class="va">sft</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSoftPower.html">getSoftPower</a></span><span class="op">(</span><span class="va">methAdj</span>, corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Soft_Power.rds"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotSoftPower.html">plotSoftPower</a></span><span class="op">(</span><span class="va">sft</span>, file <span class="op">=</span> <span class="st">"Soft_Power_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Soft_Power_Plots.png" alt=""><p class="caption">Figure 7. Soft Power Plots</p>
</div>
</div>
<div class="section level2">
<h2 id="get-comethylation-modules">Get Comethylation Modules<a class="anchor" aria-label="anchor" href="#get-comethylation-modules"></a>
</h2>
<p><code><a href="../reference/getModules.html">getModules()</a></code> identifies comethylation modules using filtered regions, a chosen soft power threshold, and either Pearson or Bicor correlation. Here we use Pearson correlation for the greater sensitivity to detect modules. Regions are first formed into blocks close to but not exceeding the maximum block size. For this particular analysis, all regions can fit into a single block. A full network analysis is then performed to assign regions to modules; modules are merged if their eigennodes are highly correlated. The modules are then saved as a .rds file.</p>
<p><code><a href="../reference/plotRegionDendro.html">plotRegionDendro()</a></code> plots region dendrograms and modules for each block. <code><a href="../reference/getModuleBED.html">getModuleBED()</a></code> creates a bed file of regions annotated with identified modules; regions in the unassigned grey module are excluded.</p>
<pre><code><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getModules.html">getModules</a></span><span class="op">(</span><span class="va">methAdj</span>, power <span class="op">=</span> <span class="va">sft</span><span class="op">$</span><span class="va">powerEstimate</span>, regions <span class="op">=</span> <span class="va">regions</span>, corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Modules.rds"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionDendro.html">plotRegionDendro</a></span><span class="op">(</span><span class="va">modules</span>, file <span class="op">=</span> <span class="st">"Region_Dendrograms.pdf"</span><span class="op">)</span>
<span class="va">BED</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getModuleBED.html">getModuleBED</a></span><span class="op">(</span><span class="va">modules</span><span class="op">$</span><span class="va">regions</span>, file <span class="op">=</span> <span class="st">"Modules.bed"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Region_Dendrograms.png" alt=""><p class="caption">Figure 8. Region Dendrograms</p>
</div>
</div>
<div class="section level2">
<h2 id="examine-correlations-between-modules-and-samples">Examine Correlations between Modules and Samples<a class="anchor" aria-label="anchor" href="#examine-correlations-between-modules-and-samples"></a>
</h2>
<p>In order to examine relationships between modules, <code>getDendro(MEs)</code> clusters modules based on eigennode values using Bicor or Pearson correlations, which are then plotted with <code><a href="../reference/plotDendro.html">plotDendro()</a></code>. <code>getCor(MEs)</code> calculates a correlation matrix for module eigennodes using Bicor or Pearson correlations, which are then plotted with <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code>. We use Bicor correlations at this stage to identify robust associations with less impact of outliers. Note that module correlation statistics (with p-values) can also be calculated with <code><a href="../reference/getMEtraitCor.html">getMEtraitCor()</a></code>. We can use these plots to identify groups of correlated modules.</p>
<pre><code><span class="va">MEs</span> <span class="op">&lt;-</span> <span class="va">modules</span><span class="op">$</span><span class="va">MEs</span>
<span class="va">moduleDendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDendro.html">getDendro</a></span><span class="op">(</span><span class="va">MEs</span>, distance <span class="op">=</span> <span class="st">"bicor"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotDendro.html">plotDendro</a></span><span class="op">(</span><span class="va">moduleDendro</span>, labelSize <span class="op">=</span> <span class="fl">5</span>, nBreaks <span class="op">=</span> <span class="fl">5</span>, file <span class="op">=</span> <span class="st">"Module_ME_Dendrogram.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Module_ME_Dendrogram.png" alt=""><p class="caption">Figure 9. Module ME Dendrogram</p>
</div>
<pre><code><span class="va">moduleCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCor.html">getCor</a></span><span class="op">(</span><span class="va">MEs</span>, corType <span class="op">=</span> <span class="st">"bicor"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">moduleCor</span>, rowDendro <span class="op">=</span> <span class="va">moduleDendro</span>, colDendro <span class="op">=</span> <span class="va">moduleDendro</span>, file <span class="op">=</span> <span class="st">"Module_Correlation_Heatmap.pdf"</span><span class="op">)</span>
<span class="va">moduleCorStats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMEtraitCor.html">getMEtraitCor</a></span><span class="op">(</span><span class="va">MEs</span>, colData <span class="op">=</span> <span class="va">MEs</span>, corType <span class="op">=</span> <span class="st">"bicor"</span>, robustY <span class="op">=</span> <span class="cn">TRUE</span>, file <span class="op">=</span> <span class="st">"Module_Correlation_Stats.txt"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Module_Correlation_Heatmap.png" alt=""><p class="caption">Figure 10. Module Correlation Heatmap</p>
</div>
<p>To explore associations between samples, <code>getDendro(MEs, transpose = TRUE)</code> clusters the samples based on module eigennode values using Bicor or Pearson correlations, which are then plotted with <code><a href="../reference/plotDendro.html">plotDendro()</a></code>. <code>getCor(MEs, transpose = TRUE)</code> calculates a correlation matrix for samples based on module eigennode vales using Bicor or Pearson correlations, which are then plotted with <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code>. <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code> can also be used to visualize module eigennode values for each sample as below. These plots can be useful to identify sets of samples with correlated methylation at these regions. This may also reveal the impact of batch effects on the identified modules.</p>
<pre><code><span class="va">sampleDendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDendro.html">getDendro</a></span><span class="op">(</span><span class="va">MEs</span>, transpose <span class="op">=</span> <span class="cn">TRUE</span>, distance <span class="op">=</span> <span class="st">"bicor"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotDendro.html">plotDendro</a></span><span class="op">(</span><span class="va">sampleDendro</span>, labelSize <span class="op">=</span> <span class="fl">3</span>, nBreaks <span class="op">=</span> <span class="fl">5</span>, file <span class="op">=</span> <span class="st">"Sample_ME_Dendrogram.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_ME_Dendrogram.png" alt=""><p class="caption">Figure 11. Sample ME Dendrogram</p>
</div>
<pre><code><span class="va">sampleCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCor.html">getCor</a></span><span class="op">(</span><span class="va">MEs</span>, transpose <span class="op">=</span> <span class="cn">TRUE</span>, corType <span class="op">=</span> <span class="st">"bicor"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">sampleCor</span>, rowDendro <span class="op">=</span> <span class="va">sampleDendro</span>, colDendro <span class="op">=</span> <span class="va">sampleDendro</span>, file <span class="op">=</span> <span class="st">"Sample_Correlation_Heatmap.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_Correlation_Heatmap.png" alt=""><p class="caption">Figure 12. Sample Correlation Heatmap</p>
</div>
<pre><code><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">MEs</span>, rowDendro <span class="op">=</span> <span class="va">sampleDendro</span>, colDendro <span class="op">=</span> <span class="va">moduleDendro</span>, legend.title <span class="op">=</span> <span class="st">"Module\nEigennode"</span>, legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.37</span>,<span class="fl">0.89</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"Sample_ME_Heatmap.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_ME_Heatmap.png" alt=""><p class="caption">Figure 13. Sample ME Heatmap</p>
</div>
</div>
<div class="section level2">
<h2 id="test-correlations-between-module-eigennodes-and-sample-traits">Test Correlations between Module Eigennodes and Sample Traits<a class="anchor" aria-label="anchor" href="#test-correlations-between-module-eigennodes-and-sample-traits"></a>
</h2>
<p>Next we can look for modules associated with sample traits. <code><a href="../reference/getMEtraitCor.html">getMEtraitCor()</a></code> tests associations between module eigennodes and sample traits using Bicor or Pearson correlation. <code><a href="../reference/getCor.html">getCor()</a></code> calculates a correlation matrix for sample traits using Bicor or Pearson correlations, which are clustered with <code><a href="../reference/getDendro.html">getDendro()</a></code>, and then plotted with <code><a href="../reference/plotDendro.html">plotDendro()</a></code>. This allows us to identify correlated traits and order both traits and modules on the heatmap by similarity. <code><a href="../reference/plotMEtraitCor.html">plotMEtraitCor()</a></code> creates a heatmap of sample traits versus modules. To focus on the top associations, another heatmap is created showing only the top 15 associations by p-value. Significant associations can be identified by a star or the p-value itself. Here we identify significant correlations between several modules and sample traits including cell type proportions, genome-wide methylation, and even home ownership.</p>
<pre><code><span class="va">MEtraitCor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMEtraitCor.html">getMEtraitCor</a></span><span class="op">(</span><span class="va">MEs</span>, colData <span class="op">=</span> <span class="va">colData</span>, corType <span class="op">=</span> <span class="st">"bicor"</span>, file <span class="op">=</span> <span class="st">"ME_Trait_Correlation_Stats.txt"</span><span class="op">)</span>
<span class="va">traitDendro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCor.html">getCor</a></span><span class="op">(</span><span class="va">MEs</span>, y <span class="op">=</span> <span class="va">colData</span>, corType <span class="op">=</span> <span class="st">"bicor"</span>, robustY <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/getDendro.html">getDendro</a></span><span class="op">(</span>transpose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotDendro.html">plotDendro</a></span><span class="op">(</span><span class="va">traitDendro</span>, labelSize <span class="op">=</span> <span class="fl">3.5</span>, expandY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.65</span>,<span class="fl">0.08</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"Trait_Dendrogram.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Trait_Dendrogram.png" alt=""><p class="caption">Figure 14. Trait Dendrogram</p>
</div>
<pre><code><span class="fu"><a href="../reference/plotMEtraitCor.html">plotMEtraitCor</a></span><span class="op">(</span><span class="va">MEtraitCor</span>, moduleOrder <span class="op">=</span> <span class="va">moduleDendro</span><span class="op">$</span><span class="va">order</span>, traitOrder <span class="op">=</span> <span class="va">traitDendro</span><span class="op">$</span><span class="va">order</span>, file <span class="op">=</span> <span class="st">"ME_Trait_Correlation_Heatmap.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/ME_Trait_Correlation_Heatmap.png" alt=""><p class="caption">Figure 15. ME Trait Correlation Heatmap</p>
</div>
<pre><code><span class="fu"><a href="../reference/plotMEtraitCor.html">plotMEtraitCor</a></span><span class="op">(</span><span class="va">MEtraitCor</span>, moduleOrder <span class="op">=</span> <span class="va">moduleDendro</span><span class="op">$</span><span class="va">order</span>, traitOrder <span class="op">=</span> <span class="va">traitDendro</span><span class="op">$</span><span class="va">order</span>, topOnly <span class="op">=</span> <span class="cn">TRUE</span>, label.type <span class="op">=</span> <span class="st">"p"</span>, label.size <span class="op">=</span> <span class="fl">4</span>, label.nudge_y <span class="op">=</span> <span class="fl">0</span>, legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.14</span>, <span class="fl">0.745</span><span class="op">)</span>, colColorMargins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">5.1</span>,<span class="fl">0.5</span>,<span class="fl">10.47</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"Top_ME_Trait_Correlation_Heatmap.pdf"</span>, width <span class="op">=</span> <span class="fl">7</span>, height <span class="op">=</span> <span class="fl">3.5</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Top_ME_Trait_Correlation_Heatmap.png" alt=""><p class="caption">Figure 16. Top ME Trait Correlation Heatmap</p>
</div>
</div>
<div class="section level2">
<h2 id="explore-significant-module-eigennode---trait-correlations">Explore Significant Module Eigennode - Trait Correlations<a class="anchor" aria-label="anchor" href="#explore-significant-module-eigennode---trait-correlations"></a>
</h2>
<div class="section level3">
<h3 id="plot-module-eigennodes-vs-traits">Plot Module Eigennodes vs Traits<a class="anchor" aria-label="anchor" href="#plot-module-eigennodes-vs-traits"></a>
</h3>
<p>To further investigate top module-trait associations, <code><a href="../reference/plotMEtraitDot.html">plotMEtraitDot()</a></code> creates a dotplot of a module eigennode by a categorical trait, while <code><a href="../reference/plotMEtraitScatter.html">plotMEtraitScatter()</a></code> creates a scatterplot of a module eigennode by a continuous trait. Any module and any sample trait can be selected. These plots can help verify that the module-trait association is robust and identify any outlier samples. Here we see that the green yellow module has lower methylation with increasing amounts of granulocytes, but higher methylation with home ownership.</p>
<pre><code><span class="fu"><a href="../reference/plotMEtraitScatter.html">plotMEtraitScatter</a></span><span class="op">(</span><span class="va">MEs</span><span class="op">$</span><span class="va">greenyellow</span>, trait <span class="op">=</span> <span class="va">colData</span><span class="op">$</span><span class="va">Gran</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.25</span>,<span class="fl">0.25</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">"Granulocytes"</span>, ylab <span class="op">=</span> <span class="st">"Green Yellow Module Eigennode"</span>, file <span class="op">=</span> <span class="st">"greenyellow_ME_Granulocytes_Scatterplot.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_ME_Granulocytes_Scatterplot.png" alt=""><p class="caption">Figure 17. Green Yellow ME Granulocytes Dotplot</p>
</div>
<pre><code><span class="fu"><a href="../reference/plotMEtraitDot.html">plotMEtraitDot</a></span><span class="op">(</span><span class="va">MEs</span><span class="op">$</span><span class="va">greenyellow</span>, trait <span class="op">=</span> <span class="va">colData</span><span class="op">$</span><span class="va">home_ownership</span>, binwidth <span class="op">=</span> <span class="fl">0.015</span>, traitCode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No"</span> <span class="op">=</span> <span class="fl">0</span>, <span class="st">"Yes"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No"</span> <span class="op">=</span> <span class="st">"#3366CC"</span>, <span class="st">"Yes"</span> <span class="op">=</span> <span class="st">"#FF3366"</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.25</span>,<span class="fl">0.25</span><span class="op">)</span>, xlab <span class="op">=</span> <span class="st">"Home Ownership"</span>, nBreaks <span class="op">=</span> <span class="fl">5</span>, ylab <span class="op">=</span> <span class="st">"Green Yellow Module Eigennode"</span>, file <span class="op">=</span> <span class="st">"greenyellow_ME_Home_Ownership_Dotplot.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_ME_Home_Ownership_Dotplot.png" alt=""><p class="caption">Figure 18. Green Yellow ME Home Ownership Dotplot</p>
</div>
</div>
<div class="section level3">
<h3 id="plot-region-methylation-vs-traits">Plot Region Methylation vs Traits<a class="anchor" aria-label="anchor" href="#plot-region-methylation-vs-traits"></a>
</h3>
<p>Next we dig further into the data and plot the raw methylation values against a sample trait using <code><a href="../reference/plotMethTrait.html">plotMethTrait()</a></code>. These values are not adjusted by principal components, just centered on the mean methylation for each region. This allows you to see the change in actual methylation across regions in a module in relation to a trait. We plot the same associations as above.</p>
<pre><code><span class="va">regions</span> <span class="op">&lt;-</span> <span class="va">modules</span><span class="op">$</span><span class="va">regions</span>
<span class="fu"><a href="../reference/plotMethTrait.html">plotMethTrait</a></span><span class="op">(</span><span class="st">"greenyellow"</span>, regions <span class="op">=</span> <span class="va">regions</span>, meth <span class="op">=</span> <span class="va">meth</span><span class="op">*</span><span class="fl">100</span>, trait <span class="op">=</span> <span class="va">colData</span><span class="op">$</span><span class="va">Gran</span>, expandY <span class="op">=</span> <span class="fl">0.04</span>, trait.legend.title <span class="op">=</span> <span class="st">"Granulocytes"</span>, trait.legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.034</span>,<span class="fl">3.35</span><span class="op">)</span>, traitMargins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">6</span>,<span class="fl">1</span>,<span class="fl">4.6</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"greenyellow_Module_Methylation_Granulocytes_Heatmap.pdf"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotMethTrait.html">plotMethTrait</a></span><span class="op">(</span><span class="st">"greenyellow"</span>, regions <span class="op">=</span> <span class="va">regions</span>, meth <span class="op">=</span> <span class="va">meth</span>, trait <span class="op">=</span> <span class="va">colData</span><span class="op">$</span><span class="va">home_ownership</span>, traitCode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No"</span> <span class="op">=</span> <span class="fl">0</span>, <span class="st">"Yes"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, traitColors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"No"</span> <span class="op">=</span> <span class="st">"#3366CC"</span>, <span class="st">"Yes"</span> <span class="op">=</span> <span class="st">"#FF3366"</span><span class="op">)</span>, trait.legend.title <span class="op">=</span> <span class="st">"Home Ownership"</span>, trait.legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.05</span>,<span class="fl">4.39</span><span class="op">)</span>, file <span class="op">=</span> <span class="st">"greenyellow_Module_Methylation_Home_Ownership_Heatmap.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_Module_Methylation_Granulocytes_Heatmap.png" alt=""><p class="caption">Figure 19. Green Yellow Module Methylation Granulocytes Heatmap</p>
</div>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_Module_Methylation_Home_Ownership_Heatmap.png" alt=""><p class="caption">Figure 20. Green Yellow Module Methylation Home Ownership Heatmap</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="annotate-modules">Annotate Modules<a class="anchor" aria-label="anchor" href="#annotate-modules"></a>
</h2>
<p><code><a href="../reference/annotateModule.html">annotateModule()</a></code> annotates a module of choice with nearby gene and CpG island context. Genes are added to regions using GREAT, gene info is added from BioMart, and both gene and CpG island context is added from <code>annotatr</code>. <code><a href="../reference/getGeneList.html">getGeneList()</a></code> then extracts just the genes for that module, in the form of the gene symbol, description, Ensembl ID, or NCBI Entrez ID.</p>
<pre><code><span class="va">regionsAnno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotateModule.html">annotateModule</a></span><span class="op">(</span><span class="va">regions</span>, module <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"greenyellow"</span><span class="op">)</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, file <span class="op">=</span> <span class="st">"Annotated_greenyellow_Module_Regions.txt"</span><span class="op">)</span>
<span class="va">geneList_greenyellow</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getGeneList.html">getGeneList</a></span><span class="op">(</span><span class="va">regionsAnno</span>, module <span class="op">=</span> <span class="st">"greenyellow"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="analyze-functional-enrichment">Analyze Functional Enrichment<a class="anchor" aria-label="anchor" href="#analyze-functional-enrichment"></a>
</h2>
<p>Next we test our module of interest for enrichment in genes with particular functions. <code><a href="../reference/listOntologies.html">listOntologies()</a></code> gets available ontologies for GREAT with the selected genome assembly. <code><a href="../reference/enrichModule.html">enrichModule()</a></code> analyzes functional enrichments for all regions assigned to the selected module, relative to the background set of all regions input into the network (including those assigned to the grey (unassigned) module). <code><a href="../reference/plotEnrichment.html">plotEnrichment()</a></code> then plots the module enrichments from GREAT.</p>
<p>Using this approach, we see that the green yellow module is enriched for genes that play a role in glutamate receptor activity and respiration.</p>
<pre><code><span class="va">ontologies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/listOntologies.html">listOntologies</a></span><span class="op">(</span><span class="st">"hg38"</span>, version <span class="op">=</span> <span class="st">"4.0.4"</span><span class="op">)</span>
<span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrichModule.html">enrichModule</a></span><span class="op">(</span><span class="va">regions</span>, module <span class="op">=</span> <span class="st">"greenyellow"</span>, genome <span class="op">=</span> <span class="st">"hg38"</span>, file <span class="op">=</span> <span class="st">"greenyellow_Module_Enrichment.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotEnrichment.html">plotEnrichment</a></span><span class="op">(</span><span class="va">enrichment</span>, file <span class="op">=</span> <span class="st">"greenyellow_Module_Enrichment_Plot.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_Module_Enrichment_Plot.png" alt=""><p class="caption">Figure 21. Green Yellow Module Enrichment Plot</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
