<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene Body Analysis • comethyl</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Gene Body Analysis">
<meta property="og:description" content="comethyl">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">comethyl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
    </li>
    <li>
      <a href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Gene_Body_Analysis_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Gene Body Analysis</h1>
                        <h4 class="author">Charles Mordaunt and Julia Mouat</h4>
            
            <h4 class="date">2021-07-12</h4>
      
      
      <div class="hidden name"><code>Gene_Body_Analysis.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Comethyl is a systems biology method for multivariate analysis of whole genome bisulfite sequencing (WGBS) data. Comethyl can be used to construct a weighted region comethylation network from user-defined regions, identify comethylation modules and the genomic regions that make up those modules, analyze functional enrichments, and investigate correlations between comethylation modules and sample traits of interest.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>You can install comethyl from this repository with the code below.</p>
<pre><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages(c("BiocManager", "remotes"))
BiocManager::install("cemordaunt/comethyl")</a></code></pre>
</div>
<div id="glossary-of-concepts" class="section level1">
<h1 class="hasAnchor">
<a href="#glossary-of-concepts" class="anchor"></a>Glossary of Concepts</h1>
<table class="table">
<colgroup>
<col width="33%">
<col width="66%">
</colgroup>
<thead><tr class="header">
<th align="left">Term</th>
<th align="left">Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Comethylation network</td>
<td align="left">A weighted network in which nodes correspond to DNA methylation and edges correspond to correlations between genes.</td>
</tr>
<tr class="even">
<td align="left">Module</td>
<td align="left">A group of genomic regions with correlated DNA methylation.</td>
</tr>
<tr class="odd">
<td align="left">Module Eigennode</td>
<td align="left">The weighted average methylation value and the first principal component of a module</td>
</tr>
<tr class="even">
<td align="left">Connectivity</td>
<td align="left">The correlation between genes in the comethylation network. The higher the connectivity, the stronger the correlation between genes.</td>
</tr>
<tr class="odd">
<td align="left">Fit</td>
<td align="left">The R-squared value of the network. This is inversely correlated to connectivity.</td>
</tr>
<tr class="even">
<td align="left">Soft Power Threshold</td>
<td align="left">The power to which all correlations are raised. This is inversely correlated to connectivity and impacts the number of modules formed.</td>
</tr>
<tr class="odd">
<td align="left">Hub gene</td>
<td align="left">The gene with the highest connectivity in a module.</td>
</tr>
</tbody>
</table>
</div>
<div id="inputs" class="section level1">
<h1 class="hasAnchor">
<a href="#inputs" class="anchor"></a>Inputs</h1>
<p><strong>Bismark Cytosine Reports:</strong> Following WGBS of your samples, the raw fastq files should be processed into CpG count matrices where biases have been removed. One available pipeline is CpG_Me (<a href="https://github.com/ben-laufer/CpG_Me" class="uri">https://github.com/ben-laufer/CpG_Me</a>), which performs read alignment and quality control of WGBS raw fastq files, and then the generates the Bismark Cytosine Reports used by Comethyl.</p>
<p><strong>Sample Trait Table</strong> (<code>"sample_info.xlsx"</code>)<strong>:</strong> This is a data frame with samples (specifying CpG reports) as rows and sample traits of interest as columns. All values in the table must be numeric, though the data can be either categorical or continuous for any given trait. Sample traits can include all available information about potential variables of interest as well as potential confounding variables. In WGBS datasets, potential confounding variables include cell type proportions as well as technical variables including coverage, read duplication, read trimming, and global cytosine methylation levels. For human populations, metadata should include clinical, diagnostic, and demographic data, as well as sample collection characteristics, such as gestational age and birthweight for cord blood. For experimental studies in animal models or cell cultures, experimental variables should be included in the metadata for exploring module-trait relationships. The comethylation modules identified from the Bismark Cytosine Reports will be correlated with the traits in this table. Keep in mind that tests for multiple hypotheses (i.e. FDR correction) will depend on the number of traits you include in your Sample Trait Table.</p>
</div>
<div id="annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation" class="anchor"></a>Annotation</h1>
<p>One unique feature of Comethyl is the ability to define regions based on functional annotations, such as CpG islands, gene bodies, enhancers, or a custom annotation. Since gene body methylation can correlate positively with expression, gene bodies were selected as alternative regions to explore in the cord blood data set, in addition to the approach of calling genomic regions by CpG location. This vignette shows the results from the gene body analysis.</p>
</div>
<div id="usage" class="section level1">
<h1 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h1>
<p>The following example is from a dataset of 74 male cord blood samples from newborns who were later diagnosed with autism spectrum disorder (ASD) and those with typical development (TD). Comethylation modules were associated with 49 sample characteristics including diagnosis, cell types, sample sequencing information such as percent CpG methylation, and demographic data such as home ownership.</p>
<div id="setup" class="section level2">
<h2 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h2>
<pre><code><a href="https://rdrr.io/r/base/library.html">library(comethyl)</a></code></pre>
</div>
<div id="set-global-options" class="section level2">
<h2 class="hasAnchor">
<a href="#set-global-options" class="anchor"></a>Set Global Options</h2>
<p><code>enableWGCNAthreads(nThreads = NULL)</code> allows multi-threading for calculations using <code>cor</code> or <code>bicor</code> functions. If the number of threads is not provided, the default number used is the number of processors online.</p>
<pre><code><a href="https://rdrr.io/r/base/options.html">options(stringsAsFactors = FALSE)
Sys.setenv(R_THREADS = 1)
enableWGCNAThreads(nThreads = 4)</a></code></pre>
</div>
<div id="read-bismark-cpg-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#read-bismark-cpg-reports" class="anchor"></a>Read Bismark CpG Reports</h2>
<p><code><a href="../reference/getCpGs.html">getCpGs()</a></code> reads individual sample Bismark CpG reports into a single BSseq object and then saves it as a .rds file. <code>colData</code> calls a data frame whose rows specify CpG reports to load into the BSseq object and whose columns are sample traits with numeric values (categorical or continuous).</p>
<pre><code>colData &lt;- read.xlsx("sample_info.xlsx", rowNames = TRUE)
bs &lt;- getCpGs(colData, file = "Unfiltered_BSseq.rds")</code></pre>
</div>
<div id="examine-cpg-totals-at-different-cutoffs" class="section level2">
<h2 class="hasAnchor">
<a href="#examine-cpg-totals-at-different-cutoffs" class="anchor"></a>Examine CpG Totals at Different Cutoffs</h2>
<p><code>getCpGtotals</code> calculates the total number and percent of CpGs remaining in a BSseq object after filtering at different <code>cov</code> (coverage) and <code>perSample</code> cutoffs and then saves it as a tab-separated text file. The purpose of this function is to help determine cutoffs to maximize the number of CpGs with sufficient data after filtering. Typically, the number of CpGs covered in 100% of samples decreases as the sample size increases, especially with low-coverage datasets. <code><a href="../reference/plotCpGtotals.html">plotCpGtotals()</a></code> plots the number of CpGs remaining after filtering by different combinations of <code>cov</code> and <code>perSample</code> in a line plot and then saves it as a PDF. <code><a href="../reference/plotCpGtotals.html">plotCpGtotals()</a></code> is designed to be used in combination with <code>getCpGtotals</code>. A ggplot is produced and can be further edited outside of this funciton if desired.</p>
<pre><code>CpGtotals &lt;- getCpGtotals(bs, file = "CpG_Totals.txt")
plotCpGtotals(CpGtotals, file = "CpG_Totals.pdf")</code></pre>
<div class="figure">
<img src="CpG%20Cluster%20Analysis/CpG_Totals.png" alt=""><p class="caption">Figure 1. CpG Totals</p>
</div>
</div>
<div id="filter-bsobject" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-bsobject" class="anchor"></a>Filter BSobject</h2>
<p><code>filterCpGs</code> subsets a BSseq object to include only those CpGs meeting <code>cov</code> and <code>perSample</code> cutoffs and then saves it as a .rds file. <code>filterCpGs</code> is designed to be used after <code>cov</code> and <code>perSample</code> arguments have been optimized by <code>getCpGtotals</code> and <code>plotCpGtotals</code>.</p>
<pre><code>bs &lt;- filterCpGs(bs, cov = 2, perSample = 0.75, file = "Filtered_BSseq.rds")</code></pre>
</div>
<div id="call-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#call-regions" class="anchor"></a>Call Regions</h2>
<p><code>getregions</code> calculates region statistics from the filtered BSobjects and saves them as a tab-separated text file. <code>getregions</code> allows for selection of annotation and genome. <code>plotRegionStats</code> plots histograms of region statistics and saves the plots as a PDF. <code>plotSDstats</code> plots methylation standard deviation versus region statistics and saves the plots as a PDF.</p>
<pre><code>regions &lt;- getRegions(bs, file = "Unfiltered_Regions.txt")
plotRegionStats(regions, maxQuantile = 0.99, file = "Unfiltered_Region_Plots.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Unfiltered_Region_Plots.png" alt=""><p class="caption">Figure 2. Unfiltered Region Plots</p>
</div>
<pre><code><a href="../reference/plotSDstats.html">plotSDstats(regions, maxQuantile = 0.99, file = "Unfiltered_SD_Plots.pdf")</a></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Unfiltered_SD_Plots.png" alt=""><p class="caption">Figure 3. Unfiltered SD Plots</p>
</div>
</div>
<div id="examine-region-totals-at-different-cutoffs" class="section level2">
<h2 class="hasAnchor">
<a href="#examine-region-totals-at-different-cutoffs" class="anchor"></a>Examine Region Totals at Different Cutoffs</h2>
<p><code>getRegionTotals</code> calculates region totals at specified covMin and methSD cutoffs and saves as a tab-separated text file. Total CpGs are expected to decrease as the minimum coverage cutoff increases and SD cutoff increases. <code>plotRegionTotals</code> plots region totals with ggplot and saves as a PDF.</p>
<pre><code>regionTotals &lt;- getRegionTotals(regions, file = "Region_Totals.txt")
plotRegionTotals(regionTotals, file = "Region_Totals.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Region_Totals.png" alt=""><p class="caption">Figure 4. Region Totals</p>
</div>
</div>
<div id="filter-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-regions" class="anchor"></a>Filter Regions</h2>
<p><code>filterRegions</code> subsets the unfiltered regions to only include those meeting <code>covMin</code> and <code>methSD</code> cutoffs and then saves it as a tab-separated text file. <code>filterRegions</code> is designed to be used after <code>covMin</code> and <code>methSD</code> functions have been optimized with <code>getRegionTotals</code> and <code>plotRegionTotals</code>. <code>plotRegionstats</code> plots total regions, total width, and total CpGs against SD cutoff, accounting for different minimum coverage cutoffs and saves them as a PDF.</p>
<pre><code>regions &lt;- filterRegions(regions, covMin = 10, methSD = 0.05, file = "Filtered_Regions.txt")
plotRegionStats(regions, maxQuantile = 0.99, file = "Filtered_Region_Plots.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Filtered_Region_Plots.png" alt=""><p class="caption">Figure 5. Filtered Region Plots</p>
</div>
</div>
<div id="adjust-methylation-data-for-principal-components" class="section level2">
<h2 class="hasAnchor">
<a href="#adjust-methylation-data-for-principal-components" class="anchor"></a>Adjust Methylation Data for Principal Components</h2>
<p><code>getRegionMeth</code> calculates region methylation from BSseq object and saves it as a .rds. <code>model.matrix</code> creates a design matrix of the region methylation. <code>adjustRegionMeth</code> adjusts the region methylation for the top principal components and saves it as a .rds file. <code>getDendro</code> clusters the samples based on the adjusted region methylation using Euclidean, Pearson, or Bicor distances. <code>plotDendro</code> creates a dendrogram of samples from the <code>getDendro</code> function and saves it as a PDF.</p>
<pre><code>meth &lt;- getRegionMeth(regions, bs = bs, file = "Region_Methylation.rds")
mod &lt;- model.matrix(~1, data = pData(bs))
methAdj &lt;- adjustRegionMeth(meth, mod = mod, file = "Adjusted_Region_Methylation.rds")
getDendro(methAdj, distance = "euclidean") %&gt;% plotDendro(file = "Sample_Dendrogram.pdf", expandY = c(0.25,0.08))</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_Dendrogram.png" alt=""><p class="caption">Figure 6. Sample Dendrogram</p>
</div>
</div>
<div id="select-soft-power-threshold" class="section level2">
<h2 class="hasAnchor">
<a href="#select-soft-power-threshold" class="anchor"></a>Select Soft Power Threshold</h2>
<p><code>getSoftPower</code> analyzes scale-free topology with either Pearson or Bicor correlations to determine the best soft-thresholding power. This refers to the power to which all correlations are raised. Pearson correlation is mean-based, assumes that the methylation data follows a normal distribution, and is sensitive to outliers. Bicor correlation is median-based and thus less sensitive to outliers, and is more robust than Pearson. <code>plotSoftPower</code> plots the soft power threshold against scale free topology fit and against connectivity. Typically, as the soft power threshold increases, fit increases and connectivity decreases. A soft power threshold should be selected where fit is 0.8 or higher.</p>
<pre><code>sft &lt;- getSoftPower(methAdj, corType = "pearson", file = "Soft_Power.rds")
plotSoftPower(sft, file = "Soft_Power_Plots.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Soft_Power_Plots.png" alt=""><p class="caption">Figure 7. Soft Power Plots</p>
</div>
</div>
<div id="get-comethylation-modules" class="section level2">
<h2 class="hasAnchor">
<a href="#get-comethylation-modules" class="anchor"></a>Get Comethylation Modules</h2>
<p><code>getModules</code> identifies comethylation modules using filtered regions, a chosen soft power threshold, and either Pearson or Bicor correlation. Regions are first formed into blocks close to but not exceeding the maximum block size. A full network analysis is then performed on each block to assign them to modules; modules are merged if they are highly correlated eigengenes. The modules are saved as a .rds file. This two-level clustering approach requires less computational memory and is significantly faster than performing full network analysis on each region. <code>plotRegionDendro</code> plots region dendrograms and modules for each block and saves as a PDF. <code>getModuleBED</code> creates a bed file of regions annotated with identified modules; regions in the unassigned grey module are excluded.</p>
<pre><code>modules &lt;- getModules(methAdj, power = sft$powerEstimate, regions = regions, corType = "pearson", file = "Modules.rds")
plotRegionDendro(modules, file = "Region_Dendrograms.pdf")
BED &lt;- getModuleBED(modules$regions, file = "Modules.bed")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Region_Dendrograms.png" alt=""><p class="caption">Figure 8. Region Dendrograms</p>
</div>
</div>
<div id="examine-correlations-between-modules-and-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#examine-correlations-between-modules-and-samples" class="anchor"></a>Examine Correlations between Modules and Samples</h2>
<p><code>moduleDendro &lt;- getDendro</code> clusters modules based on eigennode values using Bicor or Pearson correlations, which are then plotted with <code>plotDendro</code> and saved as a PDF. <code>moduleCor &lt;- getCor</code> calculates a correlation matrix for module eigennodes using Bicor or Pearson correlations, which are then plotted with <code>plotHeatmap</code> and saved as a PDF.</p>
<pre><code>MEs &lt;- modules$MEs
moduleDendro &lt;- getDendro(MEs, distance = "bicor")
plotDendro(moduleDendro, labelSize = 4, nBreaks = 5, file = "Module_ME_Dendrogram.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Module_ME_Dendrogram.png" alt=""><p class="caption">Figure 9. Module ME Dendrogram</p>
</div>
<pre><code>moduleCor &lt;- getCor(MEs, corType = "bicor")
plotHeatmap(moduleCor, rowDendro = moduleDendro, colDendro = moduleDendro, file = "Module_Correlation_Heatmap.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Module_Correlation_Heatmap.png" alt=""><p class="caption">Figure 10. Module Correlation Heatmap</p>
</div>
<p><code>sampleDendro &lt;- getDendro</code> clusters the samples based on module eignnode values using Bicor or Pearson correlations, which are then plotted with <code>plotDendro</code> and saved as a PDF. <code>sampleCor &lt;- getCor</code> calculates a correlation matrix for samples based on module eigennode vales using Bicor or Pearson correlations, which are then plotted with <code>plotHeatmap</code> and saved as a PDF.</p>
<pre><code>sampleDendro &lt;- getDendro(MEs, transpose = TRUE, distance = "bicor")
plotDendro(sampleDendro, labelSize = 3, nBreaks = 5, file = "Sample_ME_Dendrogram.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_ME_Dendrogram.png" alt=""><p class="caption">Figure 11. Sample ME Dendrogram</p>
</div>
<pre><code>sampleCor &lt;- getCor(MEs, transpose = TRUE, corType = "bicor")
plotHeatmap(sampleCor, rowDendro = sampleDendro, colDendro = sampleDendro, file = "Sample_Correlation_Heatmap.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_Correlation_Heatmap.png" alt=""><p class="caption">Figure 12. Sample Correlation Heatmap</p>
</div>
<p><code>plotHeatmap</code> plots samples against module eigennodes and saves as a PDF.</p>
<pre><code><a href="../reference/plotHeatmap.html">plotHeatmap(MEs, rowDendro = sampleDendro, colDendro = moduleDendro, legend.title = "Module\nEigennode",
            legend.position = c(0.37,0.89), file = "Sample_ME_Heatmap.pdf")</a></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sample_ME_Heatmap.png" alt=""><p class="caption">Figure 13. Sample ME Heatmap</p>
</div>
</div>
<div id="test-correlations-between-module-eigennodes-and-sample-traits" class="section level2">
<h2 class="hasAnchor">
<a href="#test-correlations-between-module-eigennodes-and-sample-traits" class="anchor"></a>Test Correlations between Module Eigennodes and Sample Traits</h2>
<p><code>getMEtraitCor</code> test associations between module eigennodes and sample traits using Bicor or Pearson correlation and the results are saved as a tab-separated text file. <code>traitDendro &lt;- getCor</code> calculates a correlation matrix for sample traits using Bicor or Pearson correlations, which are then plotted with <code>plotDendro</code> and saved as a PDF. <code>plotMEtraitCor</code> creates a heatmap of sample traits versus modules, which is saved as a PDF. Traits are ordered on the heatmap by similarity. Another heatmap is created and saved as a PDF by <code>plotMEtraitCor</code> to show only module eigennode - sample trait correlations that are significant at a specified adjusted p-value.</p>
<pre><code>MEtraitCor &lt;- getMEtraitCor(MEs, colData = colData, corType = "bicor", file = "ME_Trait_Correlation_Stats.txt")
traitDendro &lt;- getCor(MEs, y = colData, corType = "bicor", robustY = FALSE) %&gt;% getDendro(transpose = TRUE)
plotDendro(traitDendro, labelSize = 3.5, expandY = c(0.65,0.08), file = "Trait_Dendrogram.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Trait_Dendrogram.png" alt=""><p class="caption">Figure 14. Trait Dendrogram</p>
</div>
<pre><code><a href="../reference/plotMEtraitCor.html">plotMEtraitCor(MEtraitCor, moduleOrder = moduleDendro$order, traitOrder = traitDendro$order,
               file = "ME_Trait_Correlation_Heatmap.pdf")</a></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/ME_Trait_Correlation_Heatmap.png" alt=""><p class="caption">Figure 15. ME Trait Correlation Heatmap</p>
</div>
<pre><code><a href="../reference/plotMEtraitCor.html">plotMEtraitCor(MEtraitCor, moduleOrder = moduleDendro$order, traitOrder = traitDendro$order, sigOnly = TRUE, star.size = 11, star.nudge_y = -0.27, legend.position = c(1.14, 0.745), colColorMargins = c(-1,5.1,0.5,10.47), file = "Sig_ME_Trait_Correlation_Heatmap.pdf", width = 7, height = 3.5)</a></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/Sig_ME_Trait_Correlation_Heatmap.png" alt=""><p class="caption">Figure 16. Sig ME Trait Correlation Heatmap</p>
</div>
</div>
<div id="explore-significant-module-eigennode---trait-correlations" class="section level2">
<h2 class="hasAnchor">
<a href="#explore-significant-module-eigennode---trait-correlations" class="anchor"></a>Explore Significant Module Eigennode - Trait Correlations</h2>
<div id="plot-module-eigennodes-vs-traits" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-module-eigennodes-vs-traits" class="anchor"></a>Plot Module Eigennodes vs Traits</h3>
<p><code>plotMEtraitDot</code> creates dotplots of module eigennodes by categorical traits and saves as a PDF; any module and any sample trait can be selected. <code>plotMEtraitScatter</code> creates scatterplots of module eigennodes by continuous traits and saves as a PDF; any module and any sample trait can be selected.</p>
<pre><code><a href="../reference/plotMEtraitDot.html">plotMEtraitDot(MEs$bisque4, trait = colData$Diagnosis_ASD, traitCode = c("TD" = 0, "ASD" = 1),
               colors = c("TD" = "#3366CC", "ASD" = "#FF3366"), ylim = c(-0.2,0.2), xlab = "Diagnosis",
               ylab = "Bisque 4 Module Eigennode", file = "bisque4_ME_Diagnosis_Dotplot.pdf")</a></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_ME_Granulocytes_Scatterplot.png" alt=""><p class="caption">Figure 17. Green Yellow ME Granulocytes Dotplot</p>
</div>
<pre><code><a href="../reference/plotMEtraitScatter.html">plotMEtraitScatter(MEs$greenyellow, trait = colData$Gran, ylim = c(-0.15,0.15), xlab = "Granulocytes",
                   ylab = "Green Yellow Module Eigennode", file = "greenyellow_ME_Granulocytes_Scatterplot.pdf")</a></code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_ME_Home_Ownership_Dotplot.png" alt=""><p class="caption">Figure 18. Green Yellow ME Home Ownership Dotplot</p>
</div>
<pre><code><a href="../reference/plotMEtraitScatter.html">plotMEtraitScatter(MEs$greenyellow, trait = colData$home_ownership, ylim = c(-0.15,0.15), xlab = "Home Ownership",
                   ylab = "Green Yellow Module Eigennode", file = "greenyellow_ME_Home_Ownership_Scatterplot.pdf")</a></code></pre>
</div>
<div id="plot-region-methylation-vs-traits" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-region-methylation-vs-traits" class="anchor"></a>Plot Region Methylation vs Traits</h3>
<p><code>plotMethTrait</code> plots relative methylation by a sample trait and saves the heatmap as a PDF.</p>
<pre><code>regions &lt;- modules$regions
plotMethTrait("greenyellow", regions = regions, meth = meth, trait = colData$Gran, expandY = 0.04,
              trait.legend.title = "Granulocytes", trait.legend.position = c(1.034,3.35),
              file = "greenyellow_Module_Methylation_Granulocytes_Heatmap.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_Module_Methylation_Granulocytes_Heatmap.png" alt=""><p class="caption">Figure 19. Green Yellow Module Methylation Granulocytes Heatmap</p>
</div>
</div>
</div>
<div id="annotate-modules" class="section level2">
<h2 class="hasAnchor">
<a href="#annotate-modules" class="anchor"></a>Annotate Modules</h2>
<p><code>annotateModule</code> annotates a module of choice with a genome selection. Genes are added to regions using GREAT v, gene info is added from BioMart, and gene context and CpG context is added from annotatr. The file is saved as a tab-separated text file.</p>
<pre><code>regionsAnno &lt;- annotateModule(regions, module = c("greenyellow"), genome = "hg38", file = "Annotated_greenyellow_Module_Regions.txt")
geneList_greenyellow &lt;- getGeneList(regionsAnno, module = "greenyellow")
</code></pre>
</div>
<div id="analyze-functional-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#analyze-functional-enrichment" class="anchor"></a>Analyze Functional Enrichment</h2>
<p><code>listOntologies</code> gets available ontologies for GREAT v with the selected genome assembly. <code>enrichModule</code> analyzes functional enrichments for all regions assigned to the selected module and saves the result as a tab-separated text file. <code>plotEnrichment</code> plots the module enrichments from GREAT.</p>
<pre><code>ontologies &lt;- listOntologies("hg38", version = "4.0.4")
enrich_bisque4 &lt;- enrichModule(regions, module = "greenyellow", genome = "hg38", file = "greenyellow_Module_Enrichment.txt")
plotEnrichment(enrich_greenyellow, file = "greenyellow_Module_Enrichment_Plot.pdf")</code></pre>
<div class="figure">
<img src="Gene%20Body%20Analysis/greenyellow_Module_Enrichment_Plot.png" alt=""><p class="caption">Figure 20. Green Yellow Module Enrichment Plot</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
