<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="comethyl">
<title>Module Preservation • comethyl</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Module Preservation">
<meta property="og:description" content="comethyl">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">comethyl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/comethyl.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/Module_Preservation.html">Module Preservation</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cemordaunt/comethyl/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Module Preservation</h1>
                        <h4 data-toc-skip class="author">Charles Mordaunt</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cemordaunt/comethyl/blob/HEAD/vignettes/Module_Preservation.Rmd" class="external-link"><code>vignettes/Module_Preservation.Rmd</code></a></small>
      <div class="d-none name"><code>Module_Preservation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette, we use Comethyl to construct a comethylation network using clusters of CpGs grouped by genomic location in a replication dataset, and then examine the preservation of modules from the original dataset. We identify modules of comethylated regions, generate statistics of module preservation, and visualize them.</p>
<p>The original dataset explored in the <a href="https://cemordaunt.github.io/comethyl/articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a> vignette included 74 male cord blood samples from newborns who were later diagnosed with autism spectrum disorder (ASD) and those with typical development (TD). The replication dataset has the same design and includes 38 male cord blood samples. The goal of this analysis is to evaluate the consistency of our identified modules in an independent group of individuals.</p>
<p>Raw data is available on GEO (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE140730" class="external-link">GSE140730</a>), see the <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-020-00785-8" class="external-link">previous publication</a> for more details.</p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<pre><code><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cemordaunt/comethyl" class="external-link">comethyl</a></span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="set-global-options">Set Global Options<a class="anchor" aria-label="anchor" href="#set-global-options"></a>
</h2>
<p><code>WGCNA::disableWGCNAthreads()</code> prevents multi-threading for WGCNA calculations, including all functions using <code><a href="https://rdrr.io/pkg/WGCNA/man/cor.html" class="external-link">WGCNA::cor()</a></code> and <code><a href="https://rdrr.io/pkg/WGCNA/man/bicor.html" class="external-link">WGCNA::bicor()</a></code>. This is recommended for large sets of regions (&gt; 150,000). For smaller region sets, use <code>WGCNA::enableWGCNAthreads()</code> to allow for parallel calculations with the specified number of threads. If the number of threads is not provided, the default is the number of processors online.</p>
<pre><code><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>R_THREADS <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu">WGCNA</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/WGCNA/man/allowWGCNAThreads.html" class="external-link">disableWGCNAThreads</a></span><span class="op">(</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="read-bismark-cpg-reports">Read Bismark CpG Reports<a class="anchor" aria-label="anchor" href="#read-bismark-cpg-reports"></a>
</h2>
<p>In this first section, we try to match the CpGs and regions from the analysis of the first dataset to make them more comparable.</p>
<p>We read in an excel table with <code><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">openxlsx::read.xlsx()</a></code> where the first column includes the names of sample Bismark CpG reports and all other columns include trait values for each sample. All trait values must be numeric, though traits can be categorical or continuous. <code><a href="../reference/getCpGs.html">getCpGs()</a></code> reads individual sample Bismark CpG reports into a single BSseq object and then saves it as a .rds file. See <a href="https://cemordaunt.github.io/comethyl/articles/comethyl.html#inputs">Inputs</a> for more information.</p>
<p>We load the filtered BSseq object from the first analysis and use <code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">IRanges::subsetByOverlaps()</a></code> to limit our new BSseq object to those same CpGs. Then <code><a href="../reference/filterCpGs.html">filterCpGs()</a></code> subsets a BSseq object to include only those CpGs meeting <code>cov</code> and <code>perSample</code> cutoffs and then saves it as a .rds file. <code><a href="../reference/filterCpGs.html">filterCpGs()</a></code> is designed to be used after <code>cov</code> and <code>perSample</code> arguments have been optimized by <code><a href="../reference/getCpGtotals.html">getCpGtotals()</a></code> and <code><a href="../reference/plotCpGtotals.html">plotCpGtotals()</a></code>. Here we keep only CpGs with at least 1 read in at least 1 sample. We make this more inclusive than usual to better match the original dataset.</p>
<pre><code><span class="va">colData_rep</span> <span class="op">&lt;-</span> <span class="fu">openxlsx</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/openxlsx/man/read.xlsx.html" class="external-link">read.xlsx</a></span><span class="op">(</span><span class="st">"replication_sample_info.xlsx"</span>, rowNames <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">bs_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCpGs.html">getCpGs</a></span><span class="op">(</span><span class="va">colData_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Unfiltered_BSseq.rds"</span><span class="op">)</span>
<span class="va">bs_disc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Discovery_Filtered_BSseq.rds"</span><span class="op">)</span>
<span class="va">bs_rep</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">subsetByOverlaps</a></span><span class="op">(</span><span class="va">bs_rep</span>, ranges <span class="op">=</span> <span class="va">bs_disc</span><span class="op">)</span>
<span class="va">bs_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterCpGs.html">filterCpGs</a></span><span class="op">(</span><span class="va">bs_rep</span>, cov <span class="op">=</span> <span class="fl">1</span>, perSample <span class="op">=</span> <span class="fl">0.01</span>, file <span class="op">=</span> <span class="st">"Replication_Filtered_BSseq.rds"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="call-regions">Call Regions<a class="anchor" aria-label="anchor" href="#call-regions"></a>
</h2>
<p>We read in the regions from the original dataset and create a GRanges object. <code><a href="../reference/getRegions.html">getRegions()</a></code> generates a set of regions and some statistics based on the CpGs in a BSseq object. Here we use the original regions as a custom genomic annotation and input into <code><a href="../reference/getRegions.html">getRegions()</a></code>. It’s also necessary to replace <code>RegionID</code> with the <code>name</code> column so that the <code>RegionID</code>’s will match between the two datasets.</p>
<p>Next we plot the region statistics. <code><a href="../reference/plotRegionStats.html">plotRegionStats()</a></code> plots histograms of region statistics, while <code><a href="../reference/plotSDstats.html">plotSDstats()</a></code> plots methylation standard deviation versus region statistics. With these plots, we can get an idea of the characteristics of our regions and see how methylation variability is affected. The goal is to identify regions with biological variability rather than technical variability (due to low coverage).</p>
<pre><code><span class="va">regions_disc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="st">"Discovery_Filtered_Regions.txt"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
        <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">chr</span>, <span class="va">start</span>, <span class="va">end</span>, <span class="va">RegionID</span><span class="op">)</span>
<span class="va">regions_disc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">regions_disc</span>, <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="va">chr</span>, ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="va">start</span>, end <span class="op">=</span> <span class="va">end</span><span class="op">)</span>, name <span class="op">=</span> <span class="va">RegionID</span><span class="op">)</span><span class="op">)</span>
<span class="va">regions_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegions.html">getRegions</a></span><span class="op">(</span><span class="va">bs_rep</span>, custom <span class="op">=</span> <span class="va">regions_disc</span>, n <span class="op">=</span> <span class="fl">1</span>, save <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>RegionID <span class="op">=</span> <span class="va">name</span>, <span class="va">chr</span><span class="op">:</span><span class="va">methSD</span><span class="op">)</span>
<span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_tsv</a></span><span class="op">(</span><span class="va">regions_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Unfiltered_Regions.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionStats.html">plotRegionStats</a></span><span class="op">(</span><span class="va">regions_rep</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>, file <span class="op">=</span> <span class="st">"Replication_Unfiltered_Region_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Unfiltered_Region_Plots.png" alt=""><p class="caption">Figure 1. Unfiltered Region Plots</p>
</div>
<pre><code><span class="fu"><a href="../reference/plotSDstats.html">plotSDstats</a></span><span class="op">(</span><span class="va">regions_rep</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>, file <span class="op">=</span> <span class="st">"Replication_Unfiltered_SD_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Unfiltered_SD_Plots.png" alt=""><p class="caption">Figure 2. Unfiltered SD Plots</p>
</div>
</div>
<div class="section level2">
<h2 id="examine-region-totals-at-different-cutoffs">Examine Region Totals at Different Cutoffs<a class="anchor" aria-label="anchor" href="#examine-region-totals-at-different-cutoffs"></a>
</h2>
<p><code><a href="../reference/getRegionTotals.html">getRegionTotals()</a></code> calculates region totals at specified covMin and methSD cutoffs. Total regions (and thus total width and CpGs) are expected to decrease as the minimum coverage cutoff increases and SD cutoff increases. <code><a href="../reference/plotRegionTotals.html">plotRegionTotals()</a></code> plots these region totals by potential covMin and methSD cutoffs.</p>
<pre><code><span class="va">regionTotals_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegionTotals.html">getRegionTotals</a></span><span class="op">(</span><span class="va">regions_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Region_Totals.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionTotals.html">plotRegionTotals</a></span><span class="op">(</span><span class="va">regionTotals_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Region_Totals.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Region_Totals.png" alt=""><p class="caption">Figure 3. Region Totals</p>
</div>
</div>
<div class="section level2">
<h2 id="filter-regions">Filter Regions<a class="anchor" aria-label="anchor" href="#filter-regions"></a>
</h2>
<p><code><a href="../reference/filterRegions.html">filterRegions()</a></code> subsets the regions to only include those meeting <code>covMin</code> and <code>methSD</code> cutoffs. <code><a href="../reference/filterRegions.html">filterRegions()</a></code> is designed to be used after <code>covMin</code> and <code>methSD</code> functions have been optimized with <code><a href="../reference/getRegionTotals.html">getRegionTotals()</a></code> and <code><a href="../reference/plotRegionTotals.html">plotRegionTotals()</a></code>. Here we filter for regions with at least 5 reads in all samples and with a methylation standard deviation of at least 5%. Then we examine our regions again with <code><a href="../reference/plotRegionStats.html">plotRegionStats()</a></code> after filtering and continue with extracting methylation data and constructing our network.</p>
<pre><code><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterRegions.html">filterRegions</a></span><span class="op">(</span><span class="va">regions_rep</span>, covMin <span class="op">=</span> <span class="fl">5</span>, methSD <span class="op">=</span> <span class="fl">0.05</span>, file <span class="op">=</span> <span class="st">"Replication_Filtered_Regions.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotRegionStats.html">plotRegionStats</a></span><span class="op">(</span><span class="va">regions_rep</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>, file <span class="op">=</span> <span class="st">"Replication_Filtered_Region_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Filtered_Region_Plots.png" alt=""><p class="caption">Figure 4. Filtered Region Plots</p>
</div>
</div>
<div class="section level2">
<h2 id="adjust-methylation-data-for-principal-components">Adjust Methylation Data for Principal Components<a class="anchor" aria-label="anchor" href="#adjust-methylation-data-for-principal-components"></a>
</h2>
<p><code><a href="../reference/getRegionMeth.html">getRegionMeth()</a></code> calculates region methylation from a BSseq object and saves it as a .rds. <code><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix()</a></code> creates a design matrix for our set of samples. <code><a href="../reference/getPCs.html">getPCs()</a></code> calculates the top principal components, and then <code><a href="../reference/adjustRegionMeth.html">adjustRegionMeth()</a></code> adjusts the region methylation for the top PCs and saves it as a .rds file. <code><a href="../reference/getDendro.html">getDendro()</a></code> clusters the samples based on the adjusted region methylation using Euclidean distance, while <code><a href="../reference/plotDendro.html">plotDendro()</a></code> plots the dendrogram. We can use this dendrogram to see if there are any outlier samples or samples clustering separately due to batch effects.</p>
<pre><code><span class="va">meth_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegionMeth.html">getRegionMeth</a></span><span class="op">(</span><span class="va">regions_rep</span>, bs <span class="op">=</span> <span class="va">bs_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Region_Methylation.rds"</span><span class="op">)</span>
<span class="va">mod_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">1</span>, data <span class="op">=</span> <span class="fu">bsseq</span><span class="fu">::</span><span class="fu">pData</span><span class="op">(</span><span class="va">bs_rep</span><span class="op">)</span><span class="op">)</span>
<span class="va">PCs_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPCs.html">getPCs</a></span><span class="op">(</span><span class="va">meth_rep</span>, mod <span class="op">=</span> <span class="va">mod_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Top_Principal_Components.rds"</span><span class="op">)</span>
<span class="va">methAdj_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustRegionMeth.html">adjustRegionMeth</a></span><span class="op">(</span><span class="va">meth_rep</span>, PCs <span class="op">=</span> <span class="va">PCs_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Adjusted_Region_Methylation.rds"</span><span class="op">)</span>
<span class="fu"><a href="../reference/getDendro.html">getDendro</a></span><span class="op">(</span><span class="va">methAdj_rep</span>, distance <span class="op">=</span> <span class="st">"euclidean"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
<span class="fu"><a href="../reference/plotDendro.html">plotDendro</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"Replication_Sample_Dendrogram.pdf"</span>, expandY <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">0.08</span><span class="op">)</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Sample_Dendrogram.png" alt=""><p class="caption">Figure 5. Sample Dendrogram</p>
</div>
</div>
<div class="section level2">
<h2 id="select-soft-power-threshold">Select Soft Power Threshold<a class="anchor" aria-label="anchor" href="#select-soft-power-threshold"></a>
</h2>
<p><code><a href="../reference/getSoftPower.html">getSoftPower()</a></code> analyzes scale-free topology with Pearson or Bicor correlations to determine the best soft-thresholding power. This refers to the power to which all correlations are raised and how much more stronger correlations are weighted compared to weaker correlations. Pearson correlation is more sensitive than Bicor correlation, but is also more influenced by outlier samples. We use Pearson correlation in order to have higher power to detect correlated regions in a dataset with relatively low variability between samples.</p>
<p><code><a href="../reference/plotSoftPower.html">plotSoftPower()</a></code> plots the soft power threshold against scale free topology fit and connectivity. Typically, as the soft power threshold increases, fit increases and connectivity decreases. A soft power threshold should be selected as the lowest threshold where fit is 0.8 or higher. After originally trying 19, we increase the power to 25 for better fit and a mean connectivity that more closely matches the first dataset.</p>
<pre><code><span class="va">sft_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getSoftPower.html">getSoftPower</a></span><span class="op">(</span><span class="va">methAdj_rep</span>, powerVector <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Replication_Soft_Power.rds"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotSoftPower.html">plotSoftPower</a></span><span class="op">(</span><span class="va">sft_rep</span>, file <span class="op">=</span> <span class="st">"Replication_Soft_Power_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Soft_Power_Plots.png" alt=""><p class="caption">Figure 6. Soft Power Plots</p>
</div>
</div>
<div class="section level2">
<h2 id="get-comethylation-modules">Get Comethylation Modules<a class="anchor" aria-label="anchor" href="#get-comethylation-modules"></a>
</h2>
<p><code><a href="../reference/getModules.html">getModules()</a></code> identifies comethylation modules using filtered regions, a chosen soft power threshold, and either Pearson or Bicor correlation. Here we use Pearson correlation for the greater sensitivity to detect modules. Regions are first formed into blocks close to but not exceeding the maximum block size. A full network analysis is then performed on each block to assign regions to modules; modules are merged if their eigennodes are highly correlated. The modules are then saved as a .rds file. This two-level clustering approach requires less computational memory and is significantly faster than performing network analysis on all regions at once.</p>
<pre><code><span class="va">modules_rep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getModules.html">getModules</a></span><span class="op">(</span><span class="va">methAdj_rep</span>, power <span class="op">=</span> <span class="fl">25</span>, regions <span class="op">=</span> <span class="va">regions</span>, corType <span class="op">=</span> <span class="st">"pearson"</span>, nThreads <span class="op">=</span> <span class="fl">2</span>, file <span class="op">=</span> <span class="st">"Replication_Modules.rds"</span><span class="op">)</span></code></pre>
</div>
<div class="section level2">
<h2 id="evaluate-module-preservation">Evaluate Module Preservation<a class="anchor" aria-label="anchor" href="#evaluate-module-preservation"></a>
</h2>
<p>After module identification and functional association has been completed within a dataset, an important next step is the analysis of module quality and preservation in an independent replication dataset. One approach is that taken by <a href="https://doi.org/10.1371/journal.pcbi.1001057" class="external-link">Langfelder et al.</a>, where multiple statistics are calculated to assess different types of preservation based on overlapping module assignments and network connectivity features. Observed statistics are then compared with an empirical null distribution, generated by permuting module assignments 100 times, in order to calculate Z-scores and P-values. In simulation studies by Langfelder et al., Z-scores &gt; 2 were estimated as weak-to-moderate evidence for preservation, while Z-scores &gt; 10 were estimated as strong evidence.</p>
<p>To test module preservation, we load the regions with module assignments and methylation data for the original dataset and input them into <code><a href="../reference/getModulePreservation.html">getModulePreservation()</a></code> along with the new regions and associated methylation data. This calculates several preservation statistics for each module, along with Z-scores and p-values comparing them to the empirical null distribution. After waiting for the permutations to complete, we run <code><a href="../reference/plotModulePreservation.html">plotModulePreservation()</a></code> to visualize the preservation analysis. This function plots the Z-scores by module size, due to the correlation of some of the statistics with the number of regions in the module.</p>
<p>As given by the summary.qual Z-score, all modules were scored as having strong evidence for good quality, except for the Ivory module, which had weak/moderate evidence. Using the cross-tabulation derived accuracy Z-score, 40/53 modules had at least some evidence for preservation. In contrast, relatively few modules had evidence for preservation across these two datasets based on the network-derived summary.pres Z-score. Preserved modules with this statistic included the Light-Cyan, Blue and Floral-White modules. This example demonstrates the value of multiple measurements of module quality and preservation in identifying robust and reproducible comethylation modules.</p>
<pre><code><span class="va">modules_disc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Discovery_Modules.rds"</span><span class="op">)</span>
<span class="va">regions_disc</span> <span class="op">&lt;-</span> <span class="va">modules_disc</span><span class="op">$</span><span class="va">regions</span>
<span class="va">methAdj_disc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"Discovery_Adjusted_Region_Methylation.rds"</span><span class="op">)</span>
<span class="va">regions_rep</span> <span class="op">&lt;-</span> <span class="va">modules_rep</span><span class="op">$</span><span class="va">regions</span>
<span class="va">preservation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getModulePreservation.html">getModulePreservation</a></span><span class="op">(</span><span class="va">methAdj_disc</span>, regions_disc <span class="op">=</span> <span class="va">regions_disc</span>, meth_rep <span class="op">=</span> <span class="va">methAdj_rep</span>, regions_rep <span class="op">=</span> <span class="va">regions_rep</span>, corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Module_Preservation_Stats.txt"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotModulePreservation.html">plotModulePreservation</a></span><span class="op">(</span><span class="va">preservation</span>, file <span class="op">=</span> <span class="st">"Module_Preservation_Plots.pdf"</span><span class="op">)</span></code></pre>
<div class="figure">
<img src="Module%20Preservation/Module_Preservation_Statistics_Plot.png" alt=""><p class="caption">Figure 7. Module Preservation Statistic Plots</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
