<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="annotateModule() takes a data.frame of regions with module
assignments, annotates them with genes using GREAT, adds additional gene
information with Ensembl BioMart, provides gene regulatory context with
annotatr, and then saves this as a .txt file. Support is provided for several
genomes, including hg38, hg19, hg18, mm10,
mm9, and danRer7."><title>Annotate Module Regions — annotateModule • comethyl</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate Module Regions — annotateModule"><meta property="og:description" content="annotateModule() takes a data.frame of regions with module
assignments, annotates them with genes using GREAT, adds additional gene
information with Ensembl BioMart, provides gene regulatory context with
annotatr, and then saves this as a .txt file. Support is provided for several
genomes, including hg38, hg19, hg18, mm10,
mm9, and danRer7."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">comethyl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/comethyl.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Module_Preservation.html">Module Preservation</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cemordaunt/comethyl/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Annotate Module Regions</h1>
      <small class="dont-index">Source: <a href="https://github.com/cemordaunt/comethyl/blob/HEAD/R/Annotate_Modules.R" class="external-link"><code>R/Annotate_Modules.R</code></a></small>
      <div class="d-none name"><code>annotateModule.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>annotateModule()</code> takes a <code>data.frame</code> of regions with module
assignments, annotates them with genes using GREAT, adds additional gene
information with Ensembl BioMart, provides gene regulatory context with
annotatr, and then saves this as a .txt file. Support is provided for several
genomes, including <code>hg38</code>, <code>hg19</code>, <code>hg18</code>, <code>mm10</code>,
<code>mm9</code>, and <code>danRer7</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">annotateModule</span><span class="op">(</span>
  <span class="va">regions</span>,
  module <span class="op">=</span> <span class="cn">NULL</span>,
  grey <span class="op">=</span> <span class="cn">FALSE</span>,
  genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg38"</span>, <span class="st">"hg19"</span>, <span class="st">"hg18"</span>, <span class="st">"mm10"</span>, <span class="st">"mm9"</span>, <span class="st">"danRer7"</span><span class="op">)</span>,
  includeCuratedRegDoms <span class="op">=</span> <span class="cn">FALSE</span>,
  rule <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"basalPlusExt"</span>, <span class="st">"twoClosest"</span>, <span class="st">"oneClosest"</span><span class="op">)</span>,
  adv_upstream <span class="op">=</span> <span class="fl">5</span>,
  adv_downstream <span class="op">=</span> <span class="fl">1</span>,
  adv_span <span class="op">=</span> <span class="fl">1000</span>,
  adv_twoDistance <span class="op">=</span> <span class="fl">1000</span>,
  adv_oneDistance <span class="op">=</span> <span class="fl">1000</span>,
  version <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"4.0.4"</span>, <span class="st">"3.0.0"</span>, <span class="st">"2.0.2"</span><span class="op">)</span>,
  save <span class="op">=</span> <span class="cn">TRUE</span>,
  file <span class="op">=</span> <span class="st">"Annotated_Module_Regions.txt"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>regions</dt>
<dd><p>A <code>data.frame</code> of regions with module assignments,
typically obtained from <code><a href="getModules.html">getModules()</a></code>.</p></dd>
<dt>module</dt>
<dd><p>A <code>character</code> giving the name of one or more modules to
annotate. If null, all modules will be annotated.</p></dd>
<dt>grey</dt>
<dd><p>A <code>logical(1)</code> specifying whether or not to include the grey
(unassigned) module.</p></dd>
<dt>genome</dt>
<dd><p>A <code>character(1)</code> giving the genome build for the regions.
Supported genomes include <code>hg38</code>, <code>hg19</code>, <code>hg18</code>,
<code>mm10</code>, <code>mm9</code>, and <code>danRer7</code>. See <code>Details</code>.</p></dd>
<dt>includeCuratedRegDoms</dt>
<dd><p>A <code>logical(1)</code> indicating whether to include
curated regulatory domains for GREAT gene annotation.</p></dd>
<dt>rule</dt>
<dd><p>A <code>character(1)</code> specifying the rule used by GREAT for gene
annotation. Possible values include <code>basalPlusExt</code>,
<code>twoClosest</code>, and <code>oneClosest</code>. See <code><a href="https://rdrr.io/pkg/rGREAT/man/submitGreatJob.html" class="external-link">rGREAT::submitGreatJob()</a></code>
for more details for this and the next six arguments.</p></dd>
<dt>adv_upstream</dt>
<dd><p>A <code>numeric(1)</code> giving the distance upstream of the
TSS (in kb) to define a basal regulatory domain in the
<code>basalPlusExt</code> rule.</p></dd>
<dt>adv_downstream</dt>
<dd><p>A <code>numeric(1)</code> giving the distance downstream of
the TSS (in kb) to define a basal regulatory domain in the
<code>basalPlusExt</code> rule.</p></dd>
<dt>adv_span</dt>
<dd><p>A <code>numeric(1)</code> specifying the distance upstream and
downstream of the TSS (in kb) to define the maximum extension of the
regulatory domain in the <code>basalPlusExt</code> rule.</p></dd>
<dt>adv_twoDistance</dt>
<dd><p>A <code>numeric(1)</code> giving the distance upstream and
downstream of the TSS (in kb) to define the maximum extension of the
regulatory domain in the <code>twoClosest</code> rule.</p></dd>
<dt>adv_oneDistance</dt>
<dd><p>A <code>numeric(1)</code> giving the distance upstream and
downstream of the TSS (in kb) to define the maximum extension of the
regulatory domain in the <code>oneClosest</code> rule.</p></dd>
<dt>version</dt>
<dd><p>A <code>character(1)</code> specifying the version of GREAT to use
for gene annotation. Possible values include <code>4.0.4</code>,
<code>3.0.0</code>, and <code>2.0.2</code>. Different versions of GREAT support
different genomes. See <code>Details</code>.</p></dd>
<dt>save</dt>
<dd><p>A <code>logical(1)</code> indicating whether to save the
<code>data.frame</code>.</p></dd>
<dt>file</dt>
<dd><p>A <code>character(1)</code> giving the file name (.txt) for the saved
<code>data.frame</code>.</p></dd>
<dt>verbose</dt>
<dd><p>A <code>logical(1)</code> indicating whether messages should be
printed.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>data.frame</code> adding gene and regulatory annotations to the
regions.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>regions</code> can be filtered for one or more modules of interest, or the
grey (unassigned) module can be excluded. Gene annotation is performed by the
<span class="pkg">rGREAT</span> package, which allows for different annotation rules and versions
of GREAT. The default <code>basalPlusExt</code> annotation rule associates a gene
with a region if the region is within the basal regulatory domain of the gene
(5 kb upstream and 1 kb downstream of the TSS) or if it is within 1 Mb
upstream or downstream of the TSS and not in the basal regulatory domain of
another gene. Other rules include <code>twoClosest</code> and <code>oneClosest</code>,
which effectively assign the two nearest genes or one nearest genes,
respectively. See <code><a href="https://rdrr.io/pkg/rGREAT/man/submitGreatJob.html" class="external-link">rGREAT::submitGreatJob()</a></code> for more details.</p>
<p>GREAT supports different genomes depending on the version:</p><dl><dt>GREAT v4.0.4</dt>
<dd><p><code>hg38</code>, <code>hg19</code>, <code>mm10</code>, <code>mm9</code></p></dd>

<dt>GREAT v3.0.0</dt>
<dd><p><code>hg19</code>, <code>mm10</code>, <code>mm9</code>, <code>danRer7</code></p></dd>

<dt>GREAT v2.0.2</dt>
<dd><p><code>hg19</code>, <code>hg18</code>, <code>mm9</code>, <code>danRer7</code></p></dd>


</dl><p>Gene information is provided by the <span class="pkg">biomaRt</span> package, which adds the gene
description along with Ensembl and NCBI Entrez gene IDs. Regulatory context
is added by the <span class="pkg">annotatr</span> package. This provides positional context of
the region relative to nearby genes, enhancers, and CpG islands. Note that
<span class="pkg">annotatr</span> does not support the <code>hg18</code> or <code>danRer7</code> genomes.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<ul><li><p><code><a href="getModules.html">getModules()</a></code> to build a comethylation network and identify
modules of comethylated regions.</p></li>
<li><p><code><a href="getGeneList.html">getGeneList()</a></code> to extract a list of genes or IDs from the
annotated regions.</p></li>
<li><p><code><a href="listOntologies.html">listOntologies()</a></code>, <code><a href="enrichModule.html">enrichModule()</a></code>, and <code><a href="plotEnrichment.html">plotEnrichment()</a></code> to
investigate functional enrichment of module regions with GREAT.</p></li>
</ul></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get Comethylation Modules</span></span>
<span class="r-in"><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu"><a href="getModules.html">getModules</a></span><span class="op">(</span><span class="va">methAdj</span>, power <span class="op">=</span> <span class="va">sft</span><span class="op">$</span><span class="va">powerEstimate</span>, regions <span class="op">=</span> <span class="va">regions</span>,</span>
<span class="r-in">                      corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Modules.rds"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Annotate Modules</span></span>
<span class="r-in"><span class="va">regionsAnno</span> <span class="op">&lt;-</span> <span class="fu">annotateModule</span><span class="op">(</span><span class="va">regions</span>, module <span class="op">=</span> <span class="st">"bisque4"</span>,</span>
<span class="r-in">                              genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span class="r-in">                              file <span class="op">=</span> <span class="st">"Annotated_bisque4_Module_Regions.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">geneList_bisque4</span> <span class="op">&lt;-</span> <span class="fu"><a href="getGeneList.html">getGeneList</a></span><span class="op">(</span><span class="va">regionsAnno</span>, module <span class="op">=</span> <span class="st">"bisque4"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Analyze Functional Enrichment</span></span>
<span class="r-in"><span class="va">ontologies</span> <span class="op">&lt;-</span> <span class="fu"><a href="listOntologies.html">listOntologies</a></span><span class="op">(</span><span class="st">"hg38"</span>, version <span class="op">=</span> <span class="st">"4.0.4"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">enrich_bisque4</span> <span class="op">&lt;-</span> <span class="fu"><a href="enrichModule.html">enrichModule</a></span><span class="op">(</span><span class="va">regions</span>, module <span class="op">=</span> <span class="st">"bisque4"</span>,</span>
<span class="r-in">                               genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span class="r-in">                               file <span class="op">=</span> <span class="st">"bisque4_Module_Enrichment.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotEnrichment.html">plotEnrichment</a></span><span class="op">(</span><span class="va">enrich_bisque4</span>,</span>
<span class="r-in">               file <span class="op">=</span> <span class="st">"bisque4_Module_Enrichment_Plot.pdf"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

