<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Annotate Module Regions — annotateModule • comethyl</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Annotate Module Regions — annotateModule" />
<meta property="og:description" content="annotateModule() takes a data.frame of regions with module
assignments, annotates them with genes using GREAT, adds additional gene
information with Ensembl BioMart, provides gene regulatory context with
annotatr, and then saves this as a .txt file. Support is provided for several
genomes, including hg38, hg19, hg18, mm10,
mm9, and danRer7." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">comethyl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
    </li>
    <li>
      <a href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cemordaunt/comethyl/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Annotate Module Regions</h1>
    <small class="dont-index">Source: <a href='https://github.com/cemordaunt/comethyl/blob/master/R/Annotate_Modules.R'><code>R/Annotate_Modules.R</code></a></small>
    <div class="hidden name"><code>annotateModule.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>annotateModule()</code> takes a <code>data.frame</code> of regions with module
assignments, annotates them with genes using GREAT, adds additional gene
information with Ensembl BioMart, provides gene regulatory context with
annotatr, and then saves this as a .txt file. Support is provided for several
genomes, including <code>hg38</code>, <code>hg19</code>, <code>hg18</code>, <code>mm10</code>,
<code>mm9</code>, and <code>danRer7</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>annotateModule</span><span class='op'>(</span>
  <span class='va'>regions</span>,
  module <span class='op'>=</span> <span class='cn'>NULL</span>,
  grey <span class='op'>=</span> <span class='cn'>FALSE</span>,
  genome <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"hg38"</span>, <span class='st'>"hg19"</span>, <span class='st'>"hg18"</span>, <span class='st'>"mm10"</span>, <span class='st'>"mm9"</span>, <span class='st'>"danRer7"</span><span class='op'>)</span>,
  includeCuratedRegDoms <span class='op'>=</span> <span class='cn'>FALSE</span>,
  rule <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"basalPlusExt"</span>, <span class='st'>"twoClosest"</span>, <span class='st'>"oneClosest"</span><span class='op'>)</span>,
  adv_upstream <span class='op'>=</span> <span class='fl'>5</span>,
  adv_downstream <span class='op'>=</span> <span class='fl'>1</span>,
  adv_span <span class='op'>=</span> <span class='fl'>1000</span>,
  adv_twoDistance <span class='op'>=</span> <span class='fl'>1000</span>,
  adv_oneDistance <span class='op'>=</span> <span class='fl'>1000</span>,
  version <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"4.0.4"</span>, <span class='st'>"3.0.0"</span>, <span class='st'>"2.0.2"</span><span class='op'>)</span>,
  save <span class='op'>=</span> <span class='cn'>TRUE</span>,
  file <span class='op'>=</span> <span class='st'>"Annotated_Module_Regions.txt"</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>regions</th>
      <td><p>A <code>data.frame</code> of regions with module assignments, typically
obtained from <code><a href='getModules.html'>getModules()</a></code>.</p></td>
    </tr>
    <tr>
      <th>module</th>
      <td><p>A <code>character</code> giving the name of one or more modules to
annotate. If null, all modules will be annotated.</p></td>
    </tr>
    <tr>
      <th>grey</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> specifying whether or not to include the grey
(unassigned) module.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> giving the genome build for the regions.
Supported genomes include <code>hg38</code>, <code>hg19</code>, <code>hg18</code>,
<code>mm10</code>, <code>mm9</code>, and <code>danRer7</code>. See <code>Details</code>.</p></td>
    </tr>
    <tr>
      <th>includeCuratedRegDoms</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether to include
curated regulatory domains for GREAT gene annotation.</p></td>
    </tr>
    <tr>
      <th>rule</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying the rule used by GREAT for gene
annotation. Possible values include <code>basalPlusExt</code>,
<code>twoClosest</code>, and <code>oneClosest</code>. See <code><a href='https://rdrr.io/pkg/rGREAT/man/submitGreatJob.html'>rGREAT::submitGreatJob()</a></code>
for more details for this and the next six arguments.</p></td>
    </tr>
    <tr>
      <th>adv_upstream</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the distance upstream of the
TSS (in kb) to define a basal regulatory domain in the
<code>basalPlusExt</code> rule.</p></td>
    </tr>
    <tr>
      <th>adv_downstream</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the distance downstream of
the TSS (in kb) to define a basal regulatory domain in the
<code>basalPlusExt</code> rule.</p></td>
    </tr>
    <tr>
      <th>adv_span</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> specifying the distance upstream and
downstream of the TSS (in kb) to define the maximum extension of the
regulatory domain in the <code>basalPlusExt</code> rule.</p></td>
    </tr>
    <tr>
      <th>adv_twoDistance</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the distance upstream and
downstream of the TSS (in kb) to define the maximum extension of the
regulatory domain in the <code>twoClosest</code> rule.</p></td>
    </tr>
    <tr>
      <th>adv_oneDistance</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the distance upstream and
downstream of the TSS (in kb) to define the maximum extension of the
regulatory domain in the <code>oneClosest</code> rule.</p></td>
    </tr>
    <tr>
      <th>version</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying the version of GREAT to use
for gene annotation. Possible values include <code>4.0.4</code>, <code>3.0.0</code>,
and <code>2.0.2</code>. Different versions of GREAT support different
genomes. See <code>Details</code>.</p></td>
    </tr>
    <tr>
      <th>save</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether to save the
<code>data.frame</code>.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> giving the file name (.txt) for the saved
<code>data.frame</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether messages should be
printed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>data.frame</code> adding gene and regulatory annotations to the
regions.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>regions</code> can be filtered for one or more modules of interest, or the
grey (unassigned) module can be excluded. Gene annotation is performed by the
<span class="pkg">rGREAT</span> package, which allows for different annotation rules and versions
of GREAT. The default <code>basalPlusExt</code> annotation rule associates a gene
with a region if the region is within the basal regulatory domain of the gene
(5 kb upstream and 1 kb downstream of the TSS) or if it is within 1 Mb upstream
or downstream of the TSS and not in the basal regulatory domain of another gene.
Other rules include <code>twoClosest</code> and <code>oneClosest</code>, which effectively
assign the two nearest genes or one nearest genes, respectively. See
<code><a href='https://rdrr.io/pkg/rGREAT/man/submitGreatJob.html'>rGREAT::submitGreatJob()</a></code> for more details.</p>
<p>GREAT supports different genomes depending on the version:</p><dl>
<dt>GREAT v4.0.4</dt><dd><p><code>hg38</code>, <code>hg19</code>, <code>mm10</code>, <code>mm9</code></p></dd>
<dt>GREAT v3.0.0</dt><dd><p><code>hg19</code>, <code>mm10</code>, <code>mm9</code>, <code>danRer7</code></p></dd>
<dt>GREAT v2.0.2</dt><dd><p><code>hg19</code>, <code>hg18</code>, <code>mm9</code>, <code>danRer7</code></p></dd>

</dl>

<p>Gene information is provided by the <span class="pkg">biomaRt</span> package, which adds the gene
description along with Ensembl and NCBI Entrez gene IDs. Regulatory context
is added by the <span class="pkg">annotatr</span> package. This provides positional context of the
region relative to nearby genes, enhancers, and CpG islands. Note that
<span class="pkg">annotatr</span> does not support the <code>hg18</code> or <code>danRer7</code> genomes.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='getModules.html'>getModules()</a></code> to build a comethylation network and identify
modules of comethylated regions.</p></li>
<li><p><code><a href='getGeneList.html'>getGeneList()</a></code> to extract a list of genes or IDs from the
annotated regions.</p></li>
<li><p><code><a href='listOntologies.html'>listOntologies()</a></code>, <code><a href='enrichModule.html'>enrichModule()</a></code>, and <code><a href='plotEnrichment.html'>plotEnrichment()</a></code> to
investigate functional enrichment of module regions with GREAT.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>

<span class='co'># Get Comethylation Modules</span>
<span class='va'>modules</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getModules.html'>getModules</a></span><span class='op'>(</span><span class='va'>methAdj</span>, power <span class='op'>=</span> <span class='va'>sft</span><span class='op'>$</span><span class='va'>powerEstimate</span>, regions <span class='op'>=</span> <span class='va'>regions</span>,
                      corType <span class='op'>=</span> <span class='st'>"pearson"</span>, file <span class='op'>=</span> <span class='st'>"Modules.rds"</span><span class='op'>)</span>

<span class='co'># Annotate Modules</span>
<span class='va'>regionsAnno</span> <span class='op'>&lt;-</span> <span class='fu'>annotateModule</span><span class='op'>(</span><span class='va'>regions</span>, module <span class='op'>=</span> <span class='st'>"bisque4"</span>,
                              genome <span class='op'>=</span> <span class='st'>"hg38"</span>,
                              file <span class='op'>=</span> <span class='st'>"Annotated_bisque4_Module_Regions.txt"</span><span class='op'>)</span>
<span class='va'>geneList_bisque4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getGeneList.html'>getGeneList</a></span><span class='op'>(</span><span class='va'>regionsAnno</span>, module <span class='op'>=</span> <span class='st'>"bisque4"</span><span class='op'>)</span>

<span class='co'># Analyze Functional Enrichment</span>
<span class='va'>ontologies</span> <span class='op'>&lt;-</span> <span class='fu'><a href='listOntologies.html'>listOntologies</a></span><span class='op'>(</span><span class='st'>"hg38"</span>, version <span class='op'>=</span> <span class='st'>"4.0.4"</span><span class='op'>)</span>
<span class='va'>enrich_bisque4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='enrichModule.html'>enrichModule</a></span><span class='op'>(</span><span class='va'>regions</span>, module <span class='op'>=</span> <span class='st'>"bisque4"</span>,
                               genome <span class='op'>=</span> <span class='st'>"hg38"</span>,
                               file <span class='op'>=</span> <span class='st'>"bisque4_Module_Enrichment.txt"</span><span class='op'>)</span>
<span class='fu'><a href='plotEnrichment.html'>plotEnrichment</a></span><span class='op'>(</span><span class='va'>enrich_bisque4</span>,
               file <span class='op'>=</span> <span class='st'>"bisque4_Module_Enrichment_Plot.pdf"</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


