<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Identify Modules of Comethylated Regions — getModules • comethyl</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Identify Modules of Comethylated Regions — getModules" />
<meta property="og:description" content="getModules() builds a comethylation network, identifies comethylated
modules, outputs a list with region module assignments, eigennode
values, dendrograms, and module membership, and then saves this as a .rds file." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">comethyl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
    </li>
    <li>
      <a href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cemordaunt/comethyl/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify Modules of Comethylated Regions</h1>
    <small class="dont-index">Source: <a href='https://github.com/cemordaunt/comethyl/blob/master/R/Build_Comethylation_Network.R'><code>R/Build_Comethylation_Network.R</code></a></small>
    <div class="hidden name"><code>getModules.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>getModules()</code> builds a comethylation network, identifies comethylated
modules, outputs a <code>list</code> with region module assignments, eigennode
values, dendrograms, and module membership, and then saves this as a .rds file.</p>
    </div>

    <pre class="usage"><span class='fu'>getModules</span><span class='op'>(</span>
  <span class='va'>meth</span>,
  <span class='va'>power</span>,
  <span class='va'>regions</span>,
  maxBlockSize <span class='op'>=</span> <span class='fl'>40000</span>,
  corType <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"pearson"</span>, <span class='st'>"bicor"</span><span class='op'>)</span>,
  maxPOutliers <span class='op'>=</span> <span class='fl'>0.1</span>,
  deepSplit <span class='op'>=</span> <span class='fl'>4</span>,
  minModuleSize <span class='op'>=</span> <span class='fl'>10</span>,
  mergeCutHeight <span class='op'>=</span> <span class='fl'>0.1</span>,
  nThreads <span class='op'>=</span> <span class='fl'>4</span>,
  save <span class='op'>=</span> <span class='cn'>TRUE</span>,
  file <span class='op'>=</span> <span class='st'>"Modules.rds"</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>meth</th>
      <td><p>A <code>numeric matrix</code>, where each row is a sample and each
column is a region. This is typically obtained from <code><a href='adjustRegionMeth.html'>adjustRegionMeth()</a></code>.</p></td>
    </tr>
    <tr>
      <th>power</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the soft-thresholding power. This is
typically obtained from <code><a href='getSoftPower.html'>getSoftPower()</a></code>.</p></td>
    </tr>
    <tr>
      <th>regions</th>
      <td><p>A <code>data.frame</code> of regions, typically after filtering with
<code><a href='filterRegions.html'>filterRegions()</a></code>. Must have the column <code>RegionID</code>
and correspond to the regions in <code>meth</code>.</p></td>
    </tr>
    <tr>
      <th>maxBlockSize</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> specifying the maximum number of
regions in a block. If there are more than this number regions, then
regions are pre-clustered into blocks using projective K-means
clustering. Decrease this if memory is insufficient.</p></td>
    </tr>
    <tr>
      <th>corType</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> indicating which correlation statistic
to use in the adjacency calculation.</p></td>
    </tr>
    <tr>
      <th>maxPOutliers</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> specifying the maximum percentile that
can be considered outliers on each side of the median for the
<code>bicor</code> statistic.</p></td>
    </tr>
    <tr>
      <th>deepSplit</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> specifying the sensitivity for module
detection. Possible values are integers 0 to 4, with 4 having the
highest sensitivity.</p></td>
    </tr>
    <tr>
      <th>minModuleSize</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the minimum number of regions
to qualify as a module.</p></td>
    </tr>
    <tr>
      <th>mergeCutHeight</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> specifying the cut height for
merging correlated modules. Value is the maximum dissimilarity
(1 - correlation) and ranges from 0 to 1.</p></td>
    </tr>
    <tr>
      <th>nThreads</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> indicating the number of threads for
correlation calculations.</p></td>
    </tr>
    <tr>
      <th>save</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether to save the <code>list</code>.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> giving the file name (.rds) for the saved
<code>list</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether messages should be
printed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>list</code> with 11 elements. See <code><a href='https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html'>WGCNA::blockwiseModules()</a></code>
for a description of these. Additional <code>regions</code> element is a
<code>data.frame</code> with the region locations, statistics, module
assignment, module membership, and hub region status.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Comethylation networks are built and modules are identified by
<code><a href='https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html'>WGCNA::blockwiseModules()</a></code>, with <code>corType</code> set to either
<code>pearson</code> or <code>bicor</code>. Calculations are performed for a signed network
in blocks of regions of maximum size <code>maxBlockSize</code> (default = 40000).
If there are more than <code>maxBlocksize</code> regions, then regions are
pre-clustered into blocks using projective K-means clustering. Region
correlations are performed within each block and regions are clustered with
average linkage hierarchical clustering. Modules are then identified with a
dynamic hybrid tree cut and highly correlated modules are merged together.
More information is given in the documentation for <code><a href='https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html'>WGCNA::blockwiseModules()</a></code>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='getRegionMeth.html'>getRegionMeth()</a></code> and <code><a href='adjustRegionMeth.html'>adjustRegionMeth()</a></code> to extract methylation
data and then adjust it for the top principal components.</p></li>
<li><p><code><a href='getSoftPower.html'>getSoftPower()</a></code> and <code><a href='plotSoftPower.html'>plotSoftPower()</a></code> to estimate the best
soft-thresholding power and visualize scale-free topology fit
and connectivity.</p></li>
<li><p><code><a href='plotRegionDendro.html'>plotRegionDendro()</a></code> and <code><a href='getModuleBED.html'>getModuleBED()</a></code> to visualize region
similarity, genomic locations, and module assignments.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>

<span class='co'># Get Methylation Data</span>
<span class='va'>meth</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getRegionMeth.html'>getRegionMeth</a></span><span class='op'>(</span><span class='va'>regions</span>, bs <span class='op'>=</span> <span class='va'>bs</span>, file <span class='op'>=</span> <span class='st'>"Region_Methylation.rds"</span><span class='op'>)</span>

<span class='co'># Adjust Methylation Data for PCs</span>
<span class='va'>mod</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/model.matrix.html'>model.matrix</a></span><span class='op'>(</span><span class='op'>~</span><span class='fl'>1</span>, data <span class='op'>=</span> <span class='fu'>pData</span><span class='op'>(</span><span class='va'>bs</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>methAdj</span> <span class='op'>&lt;-</span> <span class='fu'><a href='adjustRegionMeth.html'>adjustRegionMeth</a></span><span class='op'>(</span><span class='va'>meth</span>, mod <span class='op'>=</span> <span class='va'>mod</span>,
                            file <span class='op'>=</span> <span class='st'>"Adjusted_Region_Methylation.rds"</span><span class='op'>)</span>

<span class='co'># Select Soft Power Threshold</span>
<span class='va'>sft</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getSoftPower.html'>getSoftPower</a></span><span class='op'>(</span><span class='va'>methAdj</span>, corType <span class='op'>=</span> <span class='st'>"pearson"</span>, file <span class='op'>=</span> <span class='st'>"Soft_Power.rds"</span><span class='op'>)</span>
<span class='fu'><a href='plotSoftPower.html'>plotSoftPower</a></span><span class='op'>(</span><span class='va'>sft</span>, file <span class='op'>=</span> <span class='st'>"Soft_Power_Plots.pdf"</span><span class='op'>)</span>

<span class='co'># Get Comethylation Modules</span>
<span class='va'>modules</span> <span class='op'>&lt;-</span> <span class='fu'>getModules</span><span class='op'>(</span><span class='va'>methAdj</span>, power <span class='op'>=</span> <span class='va'>sft</span><span class='op'>$</span><span class='va'>powerEstimate</span>, regions <span class='op'>=</span> <span class='va'>regions</span>,
                      corType <span class='op'>=</span> <span class='st'>"pearson"</span>, file <span class='op'>=</span> <span class='st'>"Modules.rds"</span><span class='op'>)</span>

<span class='co'># Visualize Comethylation Modules</span>
<span class='fu'><a href='plotRegionDendro.html'>plotRegionDendro</a></span><span class='op'>(</span><span class='va'>modules</span>, file <span class='op'>=</span> <span class='st'>"Region_Dendrograms.pdf"</span><span class='op'>)</span>
<span class='va'>BED</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getModuleBED.html'>getModuleBED</a></span><span class='op'>(</span><span class='va'>modules</span><span class='op'>$</span><span class='va'>regions</span>, file <span class='op'>=</span> <span class='st'>"Modules.bed"</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


