<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="getModules() builds a comethylation network, identifies comethylated
modules, outputs a list with region module assignments, eigennode
values, dendrograms, and module membership, and then saves this as a .rds
file."><title>Identify Modules of Comethylated Regions — getModules • comethyl</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identify Modules of Comethylated Regions — getModules"><meta property="og:description" content="getModules() builds a comethylation network, identifies comethylated
modules, outputs a list with region module assignments, eigennode
values, dendrograms, and module membership, and then saves this as a .rds
file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">comethyl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/comethyl.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Module_Preservation.html">Module Preservation</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cemordaunt/comethyl/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Identify Modules of Comethylated Regions</h1>
      <small class="dont-index">Source: <a href="https://github.com/cemordaunt/comethyl/blob/HEAD/R/Build_Comethylation_Network.R" class="external-link"><code>R/Build_Comethylation_Network.R</code></a></small>
      <div class="d-none name"><code>getModules.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>getModules()</code> builds a comethylation network, identifies comethylated
modules, outputs a <code>list</code> with region module assignments, eigennode
values, dendrograms, and module membership, and then saves this as a .rds
file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">getModules</span><span class="op">(</span>
  <span class="va">meth</span>,
  <span class="va">power</span>,
  <span class="va">regions</span>,
  maxBlockSize <span class="op">=</span> <span class="fl">40000</span>,
  corType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pearson"</span>, <span class="st">"bicor"</span><span class="op">)</span>,
  maxPOutliers <span class="op">=</span> <span class="fl">0.1</span>,
  deepSplit <span class="op">=</span> <span class="fl">4</span>,
  minModuleSize <span class="op">=</span> <span class="fl">10</span>,
  mergeCutHeight <span class="op">=</span> <span class="fl">0.1</span>,
  nThreads <span class="op">=</span> <span class="fl">4</span>,
  save <span class="op">=</span> <span class="cn">TRUE</span>,
  file <span class="op">=</span> <span class="st">"Modules.rds"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>meth</dt>
<dd><p>A <code>numeric matrix</code>, where each row is a sample and each
column is a region. This is typically obtained from
<code><a href="adjustRegionMeth.html">adjustRegionMeth()</a></code>.</p></dd>
<dt>power</dt>
<dd><p>A <code>numeric(1)</code> giving the soft-thresholding power. This is
typically obtained from <code><a href="getSoftPower.html">getSoftPower()</a></code>.</p></dd>
<dt>regions</dt>
<dd><p>A <code>data.frame</code> of regions, typically after filtering with
<code><a href="filterRegions.html">filterRegions()</a></code>. Must have the column <code>RegionID</code>
and correspond to the regions in <code>meth</code>.</p></dd>
<dt>maxBlockSize</dt>
<dd><p>A <code>numeric(1)</code> specifying the maximum number of
regions in a block. If there are more than this number regions, then
regions are pre-clustered into blocks using projective K-means
clustering. Decrease this if memory is insufficient.</p></dd>
<dt>corType</dt>
<dd><p>A <code>character(1)</code> indicating which correlation statistic
to use in the adjacency calculation.</p></dd>
<dt>maxPOutliers</dt>
<dd><p>A <code>numeric(1)</code> specifying the maximum percentile that
can be considered outliers on each side of the median for the
<code>bicor</code> statistic.</p></dd>
<dt>deepSplit</dt>
<dd><p>A <code>numeric(1)</code> specifying the sensitivity for module
detection. Possible values are integers 0 to 4, with 4 having the
highest sensitivity.</p></dd>
<dt>minModuleSize</dt>
<dd><p>A <code>numeric(1)</code> giving the minimum number of regions
to qualify as a module.</p></dd>
<dt>mergeCutHeight</dt>
<dd><p>A <code>numeric(1)</code> specifying the cut height for
merging correlated modules. Value is the maximum dissimilarity
(1 - correlation) and ranges from 0 to 1.</p></dd>
<dt>nThreads</dt>
<dd><p>A <code>numeric(1)</code> indicating the number of threads for
correlation calculations.</p></dd>
<dt>save</dt>
<dd><p>A <code>logical(1)</code> indicating whether to save the <code>list</code>.</p></dd>
<dt>file</dt>
<dd><p>A <code>character(1)</code> giving the file name (.rds) for the saved
<code>list</code>.</p></dd>
<dt>verbose</dt>
<dd><p>A <code>logical(1)</code> indicating whether messages should be
printed.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>list</code> with 11 elements. See <code><a href="https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html" class="external-link">WGCNA::blockwiseModules()</a></code>for a description of these. Additional <code>regions</code> element is a
<code>data.frame</code> with the region locations, statistics, module
assignment, module membership, and hub region status.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Comethylation networks are built and modules are identified by
<code><a href="https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html" class="external-link">WGCNA::blockwiseModules()</a></code>, with <code>corType</code> set to either
<code>pearson</code> or <code>bicor</code>. Calculations are performed for a signed
network in blocks of regions of maximum size <code>maxBlockSize</code> (default =
40000). If there are more than <code>maxBlocksize</code> regions, then regions are
pre-clustered into blocks using projective K-means clustering. Region
correlations are performed within each block and regions are clustered with
average linkage hierarchical clustering. Modules are then identified with a
dynamic hybrid tree cut and highly correlated modules are merged together.
More information is given in the documentation for <code><a href="https://rdrr.io/pkg/WGCNA/man/blockwiseModules.html" class="external-link">WGCNA::blockwiseModules()</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<ul><li><p><code><a href="getRegionMeth.html">getRegionMeth()</a></code>, <code><a href="getPCs.html">getPCs()</a></code>, and <code><a href="adjustRegionMeth.html">adjustRegionMeth()</a></code> to
extract methylation data and then adjust it for the top
principal components.</p></li>
<li><p><code><a href="getSoftPower.html">getSoftPower()</a></code> and <code><a href="plotSoftPower.html">plotSoftPower()</a></code> to estimate the best
soft-thresholding power and visualize scale-free topology fit
and connectivity.</p></li>
<li><p><code><a href="plotRegionDendro.html">plotRegionDendro()</a></code> and <code><a href="getModuleBED.html">getModuleBED()</a></code> to visualize region
similarity, genomic locations, and module assignments.</p></li>
</ul></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get Methylation Data</span></span>
<span class="r-in"><span class="va">meth</span> <span class="op">&lt;-</span> <span class="fu"><a href="getRegionMeth.html">getRegionMeth</a></span><span class="op">(</span><span class="va">regions</span>, bs <span class="op">=</span> <span class="va">bs</span>, file <span class="op">=</span> <span class="st">"Region_Methylation.rds"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Adjust Methylation Data for PCs</span></span>
<span class="r-in"><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">1</span>, data <span class="op">=</span> <span class="fu">pData</span><span class="op">(</span><span class="va">bs</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">PCs</span> <span class="op">&lt;-</span> <span class="fu"><a href="getPCs.html">getPCs</a></span><span class="op">(</span><span class="va">meth</span>, mod <span class="op">=</span> <span class="va">mod</span>, file <span class="op">=</span> <span class="st">"Top_Principal_Components.rds"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">methAdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="adjustRegionMeth.html">adjustRegionMeth</a></span><span class="op">(</span><span class="va">meth</span>, PCs <span class="op">=</span> <span class="va">PCs</span>,</span>
<span class="r-in">                            file <span class="op">=</span> <span class="st">"Adjusted_Region_Methylation.rds"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Select Soft Power Threshold</span></span>
<span class="r-in"><span class="va">sft</span> <span class="op">&lt;-</span> <span class="fu"><a href="getSoftPower.html">getSoftPower</a></span><span class="op">(</span><span class="va">methAdj</span>, corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Soft_Power.rds"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotSoftPower.html">plotSoftPower</a></span><span class="op">(</span><span class="va">sft</span>, file <span class="op">=</span> <span class="st">"Soft_Power_Plots.pdf"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Get Comethylation Modules</span></span>
<span class="r-in"><span class="va">modules</span> <span class="op">&lt;-</span> <span class="fu">getModules</span><span class="op">(</span><span class="va">methAdj</span>, power <span class="op">=</span> <span class="va">sft</span><span class="op">$</span><span class="va">powerEstimate</span>, regions <span class="op">=</span> <span class="va">regions</span>,</span>
<span class="r-in">                      corType <span class="op">=</span> <span class="st">"pearson"</span>, file <span class="op">=</span> <span class="st">"Modules.rds"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Visualize Comethylation Modules</span></span>
<span class="r-in"><span class="fu"><a href="plotRegionDendro.html">plotRegionDendro</a></span><span class="op">(</span><span class="va">modules</span>, file <span class="op">=</span> <span class="st">"Region_Dendrograms.pdf"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">BED</span> <span class="op">&lt;-</span> <span class="fu"><a href="getModuleBED.html">getModuleBED</a></span><span class="op">(</span><span class="va">modules</span><span class="op">$</span><span class="va">regions</span>, file <span class="op">=</span> <span class="st">"Modules.bed"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

