<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="getRegions() generates a set of regions and some statistics based on
the CpGs in a BSseq object and then saves it as a
tab-delimited text file. Regions can be defined based on CpG locations (for
CpG clusters), built-in genomic annotations from annotatr, or a custom
genomic annotation."><title>Generate Regions from CpGs — getRegions • comethyl</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate Regions from CpGs — getRegions"><meta property="og:description" content="getRegions() generates a set of regions and some statistics based on
the CpGs in a BSseq object and then saves it as a
tab-delimited text file. Regions can be defined based on CpG locations (for
CpG clusters), built-in genomic annotations from annotatr, or a custom
genomic annotation."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">comethyl</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/comethyl.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/Module_Preservation.html">Module Preservation</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cemordaunt/comethyl/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Generate Regions from CpGs</h1>
      <small class="dont-index">Source: <a href="https://github.com/cemordaunt/comethyl/blob/HEAD/R/Get_and_Filter_Regions.R" class="external-link"><code>R/Get_and_Filter_Regions.R</code></a></small>
      <div class="d-none name"><code>getRegions.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>getRegions()</code> generates a set of regions and some statistics based on
the CpGs in a <a href="https://rdrr.io/pkg/bsseq/man/BSseq-class.html" class="external-link">BSseq</a> object and then saves it as a
tab-delimited text file. Regions can be defined based on CpG locations (for
CpG clusters), built-in genomic annotations from <span class="pkg">annotatr</span>, or a custom
genomic annotation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">getRegions</span><span class="op">(</span>
  <span class="va">bs</span>,
  annotation <span class="op">=</span> <span class="cn">NULL</span>,
  genome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hg38"</span>, <span class="st">"hg19"</span>, <span class="st">"mm10"</span>, <span class="st">"mm9"</span>, <span class="st">"rn6"</span>, <span class="st">"rn5"</span>, <span class="st">"rn4"</span>, <span class="st">"dm6"</span>, <span class="st">"dm3"</span>,
    <span class="st">"galGal5"</span><span class="op">)</span>,
  upstream <span class="op">=</span> <span class="fl">5000</span>,
  downstream <span class="op">=</span> <span class="fl">1000</span>,
  custom <span class="op">=</span> <span class="cn">NULL</span>,
  maxGap <span class="op">=</span> <span class="fl">150</span>,
  n <span class="op">=</span> <span class="fl">3</span>,
  save <span class="op">=</span> <span class="cn">TRUE</span>,
  file <span class="op">=</span> <span class="st">"Unfiltered_Regions.txt"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>bs</dt>
<dd><p>A <a href="https://rdrr.io/pkg/bsseq/man/BSseq-class.html" class="external-link">BSseq</a> object.</p></dd>
<dt>annotation</dt>
<dd><p>A <code>character(1)</code> giving the built-in genomic
annotation to use for defining regions. Shortcuts are available for
<code>genes</code>, <code>promoters</code>, and <code>transcripts</code>. Get the
entire list of possible annotations with
<code><a href="https://rdrr.io/pkg/annotatr/man/builtin_annotations.html" class="external-link">annotatr::builtin_annotations()</a></code>, which also includes CpG islands,
enhancers, and chromatin states.</p></dd>
<dt>genome</dt>
<dd><p>A <code>character(1)</code> with the genome build to use for
built-in annotations. Available builds include <code>hg38</code>,
<code>hg19</code>, <code>mm10</code>, <code>mm9</code>, <code>rn6</code>, <code>rn5</code>,
<code>rn4</code>, <code>dm6</code>, <code>dm3</code>, and <code>galGal5</code>.</p></dd>
<dt>upstream</dt>
<dd><p>A <code>numeric(1)</code> giving the number of bases upstream of a
transcription start site to specify a promoter. Used for the
<code>promoters</code> built-in annotation.</p></dd>
<dt>downstream</dt>
<dd><p>A <code>numeric(1)</code> giving the number of bases downstream
of a transcription start site to specify a promoter. Used for the
<code>promoters</code> built-in annotation.</p></dd>
<dt>custom</dt>
<dd><p>A <a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a> object with a custom
genomic annotation for defining regions. Construct this using
<code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges()</a></code>.</p></dd>
<dt>maxGap</dt>
<dd><p>A <code>numeric(1)</code> specifying the maximum number of bases
between CpGs to be included in the same CpG cluster.</p></dd>
<dt>n</dt>
<dd><p>A <code>numeric(1)</code> giving the minimum number of CpGs for a region
to be returned. This applies to CpG clusters, built-in, and custom,
annotations.</p></dd>
<dt>save</dt>
<dd><p>A <code>logical(1)</code> indicating whether to save the
<code>data.frame</code>.</p></dd>
<dt>file</dt>
<dd><p>A <code>character(1)</code> giving the file name (.txt) for the saved
<code>data.frame</code>.</p></dd>
<dt>verbose</dt>
<dd><p>A <code>logical(1)</code> indicating whether messages should be
printed.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>data.frame</code> with the region genomic locations along with some
statistics, including number of CpGs, coverage minimum, mean, and
standard deviation, and methylation mean and standard deviation.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>These regions still need to be filtered for minimum coverage and methylation
standard deviation.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index">
<ul><li><p><code><a href="plotRegionStats.html">plotRegionStats()</a></code>, <code><a href="plotSDstats.html">plotSDstats()</a></code>, <code><a href="getRegionTotals.html">getRegionTotals()</a></code>, and
<code><a href="plotRegionTotals.html">plotRegionTotals()</a></code> for help visualizing region
characteristics and setting cutoffs for filtering.</p></li>
<li><p><code><a href="filterRegions.html">filterRegions()</a></code> for filtering regions by minimum coverage and
methylation standard deviation.</p></li>
</ul></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Call Regions</span></span>
<span class="r-in"><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">getRegions</span><span class="op">(</span><span class="va">bs</span>, file <span class="op">=</span> <span class="st">"Unfiltered_Regions.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotRegionStats.html">plotRegionStats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span class="r-in">                file <span class="op">=</span> <span class="st">"Unfiltered_Region_Plots.pdf"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotSDstats.html">plotSDstats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span class="r-in">            file <span class="op">=</span> <span class="st">"Unfiltered_SD_Plots.pdf"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Examine Region Totals at Different Cutoffs</span></span>
<span class="r-in"><span class="va">regionTotals</span> <span class="op">&lt;-</span> <span class="fu"><a href="getRegionTotals.html">getRegionTotals</a></span><span class="op">(</span><span class="va">regions</span>, file <span class="op">=</span> <span class="st">"Region_Totals.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotRegionTotals.html">plotRegionTotals</a></span><span class="op">(</span><span class="va">regionTotals</span>, file <span class="op">=</span> <span class="st">"Region_Totals.pdf"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Filter Regions</span></span>
<span class="r-in"><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="filterRegions.html">filterRegions</a></span><span class="op">(</span><span class="va">regions</span>, covMin <span class="op">=</span> <span class="fl">10</span>, methSD <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span class="r-in">                         file <span class="op">=</span> <span class="st">"Filtered_Regions.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotRegionStats.html">plotRegionStats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span class="r-in">                file <span class="op">=</span> <span class="st">"Filtered_Region_Plots.pdf"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="plotSDstats.html">plotSDstats</a></span><span class="op">(</span><span class="va">regions</span>, maxQuantile <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span class="r-in">            file <span class="op">=</span> <span class="st">"Filtered_SD_Plots.pdf"</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

