<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate Regions from CpGs — getRegions • comethyl</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Generate Regions from CpGs — getRegions" />
<meta property="og:description" content="getRegions() generates a set of regions and some statistics based on
the CpGs in a BSseq object and then saves it as a
tab-delimited text file. Regions can be defined based on CpG locations (for
CpG clusters), built-in genomic annotations from annotatr, or a custom
genomic annotation." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">comethyl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/CpG_Cluster_Analysis.html">CpG Cluster Analysis</a>
    </li>
    <li>
      <a href="../articles/Gene_Body_Analysis.html">Gene Body Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate Regions from CpGs</h1>
    
    <div class="hidden name"><code>getRegions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>getRegions()</code> generates a set of regions and some statistics based on
the CpGs in a <a href='https://rdrr.io/pkg/bsseq/man/BSseq-class.html'>BSseq</a> object and then saves it as a
tab-delimited text file. Regions can be defined based on CpG locations (for
CpG clusters), built-in genomic annotations from <span class="pkg">annotatr</span>, or a custom
genomic annotation.</p>
    </div>

    <pre class="usage"><span class='fu'>getRegions</span><span class='op'>(</span>
  <span class='va'>bs</span>,
  annotation <span class='op'>=</span> <span class='cn'>NULL</span>,
  genome <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"hg38"</span>, <span class='st'>"hg19"</span>, <span class='st'>"mm10"</span>, <span class='st'>"mm9"</span>, <span class='st'>"rn6"</span>, <span class='st'>"rn5"</span>, <span class='st'>"rn4"</span>, <span class='st'>"dm6"</span>, <span class='st'>"dm3"</span>,
    <span class='st'>"galGal5"</span><span class='op'>)</span>,
  upstream <span class='op'>=</span> <span class='fl'>5000</span>,
  downstream <span class='op'>=</span> <span class='fl'>1000</span>,
  custom <span class='op'>=</span> <span class='cn'>NULL</span>,
  maxGap <span class='op'>=</span> <span class='fl'>150</span>,
  n <span class='op'>=</span> <span class='fl'>3</span>,
  save <span class='op'>=</span> <span class='cn'>TRUE</span>,
  file <span class='op'>=</span> <span class='st'>"Unfiltered_Regions.txt"</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bs</th>
      <td><p>A <a href='https://rdrr.io/pkg/bsseq/man/BSseq-class.html'>BSseq</a> object.</p></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> giving the built-in genomic
annotation to use for defining regions. Shortcuts are available for
<code>genes</code>, <code>promoters</code>, and <code>transcripts</code>. Get the
entire list of possible annotations with
<code><a href='https://rdrr.io/pkg/annotatr/man/builtin_annotations.html'>annotatr::builtin_annotations()</a></code>, which also includes CpG islands,
enhancers, and chromatin states.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> with the genome build to use for
built-in annotations. Available builds include <code>hg38</code>,
<code>hg19</code>, <code>mm10</code>, <code>mm9</code>, <code>rn6</code>, <code>rn5</code>,
<code>rn4</code>, <code>dm6</code>, <code>dm3</code>, and <code>galGal5</code>.</p></td>
    </tr>
    <tr>
      <th>upstream</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the number of bases upstream of a
transcription start site to specify a promoter. Used for the
<code>promoters</code> built-in annotation.</p></td>
    </tr>
    <tr>
      <th>downstream</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the number of bases downstream
of a transcription start site to specify a promoter. Used for the
<code>promoters</code> built-in annotation.</p></td>
    </tr>
    <tr>
      <th>custom</th>
      <td><p>A <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a> object with a custom
genomic annotation for defining regions. Construct this using
<code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GenomicRanges::GRanges()</a></code>.</p></td>
    </tr>
    <tr>
      <th>maxGap</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> specifying the maximum number of bases
between CpGs to be included in the same CpG cluster.</p></td>
    </tr>
    <tr>
      <th>n</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/numeric.html'>numeric(1)</a></code> giving the minimum number of CpGs for a region
to be returned. This applies to CpG clusters, built-in, and custom,
annotations.</p></td>
    </tr>
    <tr>
      <th>save</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether to save the
<code>data.frame</code>.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> giving the file name (.txt) for the saved
<code>data.frame</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code> indicating whether messages should be
printed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>data.frame</code> with the region genomic locations along with some
statistics, including number of CpGs, coverage minimum, mean, and
standard deviation, and methylation mean and standard deviation.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>These regions still need to be filtered for minimum coverage and methylation
standard deviation.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='plotRegionStats.html'>plotRegionStats()</a></code>, <code><a href='plotSDstats.html'>plotSDstats()</a></code>, <code><a href='getRegionTotals.html'>getRegionTotals()</a></code>, and
<code><a href='plotRegionTotals.html'>plotRegionTotals()</a></code> for help visualizing region
characteristics and setting cutoffs for filtering.</p></li>
<li><p><code><a href='filterRegions.html'>filterRegions()</a></code> for filtering regions by minimum coverage and
methylation standard deviation.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>

<span class='co'># Call Regions</span>
<span class='va'>regions</span> <span class='op'>&lt;-</span> <span class='fu'>getRegions</span><span class='op'>(</span><span class='va'>bs</span>, file <span class='op'>=</span> <span class='st'>"Unfiltered_Regions.txt"</span><span class='op'>)</span>
<span class='fu'><a href='plotRegionStats.html'>plotRegionStats</a></span><span class='op'>(</span><span class='va'>regions</span>, maxQuantile <span class='op'>=</span> <span class='fl'>0.99</span>,
                file <span class='op'>=</span> <span class='st'>"Unfiltered_Region_Plots.pdf"</span><span class='op'>)</span>
<span class='fu'><a href='plotSDstats.html'>plotSDstats</a></span><span class='op'>(</span><span class='va'>regions</span>, maxQuantile <span class='op'>=</span> <span class='fl'>0.99</span>,
            file <span class='op'>=</span> <span class='st'>"Unfiltered_SD_Plots.pdf"</span><span class='op'>)</span>

<span class='co'># Examine Region Totals at Different Cutoffs</span>
<span class='va'>regionTotals</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getRegionTotals.html'>getRegionTotals</a></span><span class='op'>(</span><span class='va'>regions</span>, file <span class='op'>=</span> <span class='st'>"Region_Totals.txt"</span><span class='op'>)</span>
<span class='fu'><a href='plotRegionTotals.html'>plotRegionTotals</a></span><span class='op'>(</span><span class='va'>regionTotals</span>, file <span class='op'>=</span> <span class='st'>"Region_Totals.pdf"</span><span class='op'>)</span>

<span class='co'># Filter Regions</span>
<span class='va'>regions</span> <span class='op'>&lt;-</span> <span class='fu'><a href='filterRegions.html'>filterRegions</a></span><span class='op'>(</span><span class='va'>regions</span>, covMin <span class='op'>=</span> <span class='fl'>10</span>, methSD <span class='op'>=</span> <span class='fl'>0.05</span>,
                         file <span class='op'>=</span> <span class='st'>"Filtered_Regions.txt"</span><span class='op'>)</span>
<span class='fu'><a href='plotRegionStats.html'>plotRegionStats</a></span><span class='op'>(</span><span class='va'>regions</span>, maxQuantile <span class='op'>=</span> <span class='fl'>0.99</span>,
                file <span class='op'>=</span> <span class='st'>"Filtered_Region_Plots.pdf"</span><span class='op'>)</span>
<span class='fu'><a href='plotSDstats.html'>plotSDstats</a></span><span class='op'>(</span><span class='va'>regions</span>, maxQuantile <span class='op'>=</span> <span class='fl'>0.99</span>,
            file <span class='op'>=</span> <span class='st'>"Filtered_SD_Plots.pdf"</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Charles Mordaunt.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


